{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "16145df8",
   "metadata": {},
   "source": [
    "# Immune Receptor Preprocessing\n",
    "\n",
    "### Chapter Goals:\n",
    "- Understanding output of allignment pipelines\n",
    "- Allingning into cell-based format\n",
    "- Asssignment of IR state\n",
    "\n",
    "\n",
    "### Content:\n",
    "1. Intro Immune Receptor Profiling\n",
    "2. Dataset\n",
    "3. Loading Data\n",
    "    1. Understanding Raw Data\n",
    "    2. Sort to cell format\n",
    "4. Assign IR state\n",
    "    1. Single Chains\n",
    "    2. Multiple IRs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77939925",
   "metadata": {},
   "source": [
    "## 1. Introduction: Immune Receptor Profiling\n",
    "- Role IR in cells => highly polymorphic immune receptors (IR) in from of TCR and BCR\n",
    "    - TCR interacts with pMHC to detect non-self\n",
    "    - BCR binds to free stuff => produces Antibodies later on\n",
    "- Composition of IR via VDJ\n",
    "    - two chains (TCR: majoriy alpha/beta some gamma/delta (innate like) (BCR: \n",
    "    - Variable diversity joining segments\n",
    "    - VJ = light chain, TRA\n",
    "    - VDJ = heavy chain, TRB\n",
    "    - Formation via recombination process => combinatorial diversity + junctional (?) diversity (TCR)\n",
    "    - Large diversity => detection of various disease / autoimmunity \n",
    "    - Complmentary Determining region 1-3 most variable => CDR3 main for binding\n",
    "- Assay strategy?\n",
    "    - single cell allows joined alpha beta IRs\n",
    "    - Gel Bead in Emulsion + Barcoding\n",
    "    - Targeted enrichment "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca310fcf",
   "metadata": {},
   "source": [
    "## 2. Dataset\n",
    "\n",
    "To showcase different approaches for preprocessing and analysis of IRs, we will use the dataset published by the Haniffa Lab (Stepenson et al, [1]), which contains transcriptome sequencing for over 750,000 Peripheral Blood Mononuclear Cells (PBMC) from 130 patients in the context of Severe Acute Respiratory Syndrome Coronavirus 2 (SARS-CoV-2).\n",
    "\n",
    "It consists of patients from three sources (Newcastle, Cambridge, and London) with different degrees of severity (assymptomatic, mild, mederate, severe, and critical) as well as negative controls (healthy, other severe respiratory illnesses, and healthy with Intravenous Liposaccharide to mimic systemic inflammatory response). Additional, patient-level information such as age span, gender, and smoker status are provided. \n",
    "\n",
    "The dataset was choosen as an expample of a large-scale single cell dataset including VDJ-Sequencing, that is well known to the community. It contains over 150,000 B-Cells and 200,000 T-Cells with IR annotation, that we will regard in this analysis tutorial."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "140087f0",
   "metadata": {},
   "source": [
    "To download the data please run the following commands:\n",
    "\n",
    "```\n",
    "wget https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-10026/E-MTAB-10026.processed.1.zip\n",
    "unzip E-MTAB-10026.processed.1.zip\n",
    "\n",
    "wget https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-10026/E-MTAB-10026.processed.2.zip\n",
    "unzip E-MTAB-10026.processed.2.zip\n",
    "```\n",
    "\n",
    "Or download and unzip from:\n",
    "- https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-10026/E-MTAB-10026.processed.1.zip\n",
    "- https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-10026/E-MTAB-10026.processed.2.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5923916",
   "metadata": {},
   "source": [
    "[1] Stephenson, E., Reynolds, G., Botting, R.A. et al. Single-cell multi-omics analysis of the immune response in COVID-19. Nat Med 27, 904–916 (2021). https://doi.org/10.1038/s41591-021-01329-2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec593916",
   "metadata": {},
   "source": [
    "## 3. Load data\n",
    "In this tutorial we will mainly use 2 python packages for loading, allignment?, and visualiation\n",
    "- Scanpy: general package for single cell analysis (https://github.com/theislab/scanpy, [2])\n",
    "- Scirpy: scanpy extension for immune receptor analysis (https://github.com/scverse/scirpy, [3])\n",
    "\n",
    "Further, we use additional python packages commonly used for data handling, processing, and visualisation.\n",
    "\n",
    "<br />\n",
    "<br />\n",
    "[2] Wolf, F., Angerer, P. & Theis, F. SCANPY: large-scale single-cell gene expression data analysis. Genome Biol 19, 15 (2018). https://doi.org/10.1186/s13059-017-1382-0\n",
    "\n",
    "[3] Gregor Sturm, Tamas Szabo, Georgios Fotakis, Marlene Haider, Dietmar Rieder, Zlatko Trajanoski, Francesca Finotello, Scirpy: a Scanpy extension for analyzing single-cell T-cell receptor-sequencing data, Bioinformatics, Volume 36, Issue 18, 15 September 2020, Pages 4817–4818, https://doi.org/10.1093/bioinformatics/btaa611"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1001553b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\felix.drost\\AppData\\Roaming\\Python\\Python37\\site-packages\\statsmodels\\tools\\_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
      "  import pandas.util.testing as tm\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import scirpy as ir\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b27fd49",
   "metadata": {},
   "source": [
    "Let's set the input and output paths of our data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7567d8b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_data = '../data'\n",
    "\n",
    "path_bcr_input = f'{path_data}/scbcr_cellranger_filtered_contig_annotations.csv.gz'\n",
    "path_tcr_input = f'{path_data}/TCR_merged-Updated.tsv'\n",
    "path_tcr_csv = f'{path_data}/TCR_merged-Updated.csv'\n",
    "path_tcr_out = '...'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c252c94",
   "metadata": {},
   "source": [
    "### 3.1. Raw data\n",
    "We beginn the analysis with viewing the raw output of the cell ranger pipeline for a better understanding of the data we are working with.\n",
    "We will load the \"filtered_contig_annotations.csv\"-File to view its content. Each row will represent one measurement of a sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cbf62761",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total measurements:  373670\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\felix.drost\\Anaconda3\\envs\\covidVac\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3186: DtypeWarning: Columns (12,20) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>newBarcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGCGATATA-MH9143270</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGCGATATA-MH9143270_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>646</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV1-44</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ1</td>\n",
       "      <td>IGLC2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CAAWDDSLSASYVF</td>\n",
       "      <td>TGTGCAGCATGGGATGACAGCCTGAGTGCCTCTTATGTCTTC</td>\n",
       "      <td>154</td>\n",
       "      <td>14</td>\n",
       "      <td>clonotype9</td>\n",
       "      <td>clonotype9_consensus_2</td>\n",
       "      <td>COVID-002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGCGATATA-MH9143270</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGCGATATA-MH9143270_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>579</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-33</td>\n",
       "      <td>IGHD3-3</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARDNRVYYDFWSGYPDYW</td>\n",
       "      <td>TGTGCGAGAGACAACCGAGTCTATTACGATTTTTGGAGTGGTTATC...</td>\n",
       "      <td>144</td>\n",
       "      <td>13</td>\n",
       "      <td>clonotype9</td>\n",
       "      <td>clonotype9_consensus_1</td>\n",
       "      <td>COVID-002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGTCTACCAGA-MH9143270</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGTCTACCAGA-MH9143270_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>574</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV1-69D</td>\n",
       "      <td>IGHD2-15</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARDRRSAYCSGGSCWGGDWFDPW</td>\n",
       "      <td>TGTGCGAGAGATCGTCGTTCAGCTTATTGTAGTGGTGGTAGCTGCT...</td>\n",
       "      <td>564</td>\n",
       "      <td>42</td>\n",
       "      <td>clonotype10</td>\n",
       "      <td>clonotype10_consensus_1</td>\n",
       "      <td>COVID-002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGTCTACCAGA-MH9143270</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGTCTACCAGA-MH9143270_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>744</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV1-51</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ3</td>\n",
       "      <td>IGLC3</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CGTWDSSLSAWVF</td>\n",
       "      <td>TGCGGAACATGGGATAGCAGCCTGAGTGCTTGGGTGTTC</td>\n",
       "      <td>188</td>\n",
       "      <td>21</td>\n",
       "      <td>clonotype10</td>\n",
       "      <td>None</td>\n",
       "      <td>COVID-002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGGGCATTTGCTT-MH9143270</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGCATTTGCTT-MH9143270_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>565</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-33</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CANPTRPYSSSWWYFDYW</td>\n",
       "      <td>TGTGCGAACCCCACCCGTCCATATAGCAGCAGCTGGTGGTACTTTG...</td>\n",
       "      <td>234</td>\n",
       "      <td>20</td>\n",
       "      <td>clonotype11</td>\n",
       "      <td>clonotype11_consensus_1</td>\n",
       "      <td>COVID-002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      barcode  is_cell                            contig_id  \\\n",
       "0  AAACCTGAGCGATATA-MH9143270     True  AAACCTGAGCGATATA-MH9143270_contig_1   \n",
       "1  AAACCTGAGCGATATA-MH9143270     True  AAACCTGAGCGATATA-MH9143270_contig_2   \n",
       "2  AAACCTGTCTACCAGA-MH9143270     True  AAACCTGTCTACCAGA-MH9143270_contig_1   \n",
       "3  AAACCTGTCTACCAGA-MH9143270     True  AAACCTGTCTACCAGA-MH9143270_contig_2   \n",
       "4  AAACGGGCATTTGCTT-MH9143270     True  AAACGGGCATTTGCTT-MH9143270_contig_1   \n",
       "\n",
       "   high_confidence  length chain     v_gene    d_gene j_gene c_gene  \\\n",
       "0             True     646   IGL   IGLV1-44      None  IGLJ1  IGLC2   \n",
       "1             True     579   IGH   IGHV3-33   IGHD3-3  IGHJ4   IGHM   \n",
       "2             True     574   IGH  IGHV1-69D  IGHD2-15  IGHJ5   IGHM   \n",
       "3             True     744   IGL   IGLV1-51      None  IGLJ3  IGLC3   \n",
       "4             True     565   IGH   IGHV3-33      None  IGHJ4   IGHM   \n",
       "\n",
       "   full_length productive                      cdr3  \\\n",
       "0         True       True            CAAWDDSLSASYVF   \n",
       "1         True       True       CARDNRVYYDFWSGYPDYW   \n",
       "2         True       True  CARDRRSAYCSGGSCWGGDWFDPW   \n",
       "3         True      False             CGTWDSSLSAWVF   \n",
       "4         True       True        CANPTRPYSSSWWYFDYW   \n",
       "\n",
       "                                             cdr3_nt  reads  umis  \\\n",
       "0         TGTGCAGCATGGGATGACAGCCTGAGTGCCTCTTATGTCTTC    154    14   \n",
       "1  TGTGCGAGAGACAACCGAGTCTATTACGATTTTTGGAGTGGTTATC...    144    13   \n",
       "2  TGTGCGAGAGATCGTCGTTCAGCTTATTGTAGTGGTGGTAGCTGCT...    564    42   \n",
       "3            TGCGGAACATGGGATAGCAGCCTGAGTGCTTGGGTGTTC    188    21   \n",
       "4  TGTGCGAACCCCACCCGTCCATATAGCAGCAGCTGGTGGTACTTTG...    234    20   \n",
       "\n",
       "  raw_clonotype_id         raw_consensus_id patient_id newBarcode  \n",
       "0       clonotype9   clonotype9_consensus_2  COVID-002        NaN  \n",
       "1       clonotype9   clonotype9_consensus_1  COVID-002        NaN  \n",
       "2      clonotype10  clonotype10_consensus_1  COVID-002        NaN  \n",
       "3      clonotype10                     None  COVID-002        NaN  \n",
       "4      clonotype11  clonotype11_consensus_1  COVID-002        NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bcr_raw = pd.read_csv(path_bcr_input, index_col=0)\n",
    "# 'productive' loaded with mixed type (boolean and string) => correct\n",
    "df_bcr_raw['productive'] = df_bcr_raw['productive'].astype(str)\n",
    "print('Total measurements: ', len(df_bcr_raw))\n",
    "df_bcr_raw.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "971c911e",
   "metadata": {},
   "source": [
    "The Table contains the following entries, that will be relevant for our preprocessing or analysis. A more detailed explanation of cellranger specific entries can be found here (https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation). In other data formats entries (e.g. AIRR) will have similiar, deviating names, however the underlying information remains similiar.\n",
    "- barcode: tag of the cell the contig was measured from\n",
    "- is_cell: indicates whether the barcode is associated with a cell\n",
    "- high_confidence: confidence of the measurement being a IR\n",
    "- chain: chain of the IR (e.g. TRA: T Cell Receptor alpha chain, IGH: Immuneglobulin Heavy chain)\n",
    "- {v,d,j,c}_gene: gene used to form the specific segment of the IR\n",
    "- full_length: whether the full IR was captured (see below)\n",
    "- productive: whether the IR is productive (see below)\n",
    "- cdr3{_nt}: {Nucleotide} sequence of the Complementary determining region 3 of the IR chain\n",
    "- patient_id: ID indicating the different patients\n",
    "\n",
    "(reads, umis => explain them?)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "934d0ede",
   "metadata": {},
   "source": [
    "#### Productive IRs\n",
    "Todo: General blabla on IRs productive vs non productive\n",
    "Productive Immune receptors are defined as:\n",
    "(TODO: 10x definition (https://kb.10xgenomics.com/hc/en-us/articles/115003248383-What-are-productive-contigs-) => second source)\n",
    "- Sequences spanning over from a V gene to a J-gene\n",
    "- Having a start codon in the leading region\n",
    "- Containing a CDR3 in the frame of the start codon.\n",
    "- Do not contain a stop codon within the V-J span\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "479a0b06",
   "metadata": {},
   "source": [
    "Example 1: Contigs not spanning over a full IR (V to J gene) as indicated by the 'full_length' column. You can find missing gene annotation here especially for V and J genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "45bce75a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>productive</th>\n",
       "      <th>full_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>AACTCCCCAAGTACCT-MH9143270</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLC3</td>\n",
       "      <td>IGLJ5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>AACTCTTCAGATCGGA-MH9143270</td>\n",
       "      <td>IGKV1-39</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>AACTGGTCAGGGAGAG-MH9143270</td>\n",
       "      <td>IGLV4-3</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLC2</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>ACACCAAGTCGATTGT-MH9143270</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHD5-12</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ACATCAGCATACTACG-MH9143270</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHD3-3</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       barcode    v_gene    d_gene c_gene j_gene productive  \\\n",
       "34  AACTCCCCAAGTACCT-MH9143270      None      None  IGLC3  IGLJ5      False   \n",
       "38  AACTCTTCAGATCGGA-MH9143270  IGKV1-39      None   IGKC   None      False   \n",
       "43  AACTGGTCAGGGAGAG-MH9143270   IGLV4-3      None  IGLC2   None      False   \n",
       "65  ACACCAAGTCGATTGT-MH9143270      None  IGHD5-12   IGHD  IGHJ4      False   \n",
       "76  ACATCAGCATACTACG-MH9143270      None   IGHD3-3   IGHD  IGHJ6      False   \n",
       "\n",
       "    full_length  \n",
       "34        False  \n",
       "38        False  \n",
       "43        False  \n",
       "65        False  \n",
       "76        False  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = ['barcode', 'v_gene', 'd_gene', 'c_gene', 'j_gene', 'productive', 'full_length']\n",
    "df_bcr_raw[~df_bcr_raw['full_length']][columns].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1459ec7e",
   "metadata": {},
   "source": [
    "Example 2: Contigs express full length but there is not identifieable CDR3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8c06daec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>productive</th>\n",
       "      <th>full_length</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGTCTACCAGA-MH9143270</td>\n",
       "      <td>IGLV1-51</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLC3</td>\n",
       "      <td>IGLJ3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>CGTWDSSLSAWVF</td>\n",
       "      <td>TGCGGAACATGGGATAGCAGCCTGAGTGCTTGGGTGTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AAAGCAAGTCGAAAGC-MH9143270</td>\n",
       "      <td>IGHV5-51</td>\n",
       "      <td>IGHD2-15</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AACGTTGAGCTGCGAA-MH9143270</td>\n",
       "      <td>IGHV3-30</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>CESGRATITMIVVTDLDYW</td>\n",
       "      <td>TGCGAAAGCGGGCGGGCTACGATTACTATGATAGTAGTGACTGACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AACTCAGTCAAAGTAG-MH9143270</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>IGHD3-16</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>CAKAWIGLDSEIYYDYIWGSYRIDFDYW</td>\n",
       "      <td>TGTGCGAAAGCTTGGATCGGACTCGACTCCGAGATTTATTATGATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>AACTCCCAGGTGCACA-MH9143270</td>\n",
       "      <td>IGKV6-21</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>IGKJ3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       barcode    v_gene    d_gene c_gene j_gene productive  \\\n",
       "3   AAACCTGTCTACCAGA-MH9143270  IGLV1-51      None  IGLC3  IGLJ3      False   \n",
       "13  AAAGCAAGTCGAAAGC-MH9143270  IGHV5-51  IGHD2-15   IGHD  IGHJ4      False   \n",
       "23  AACGTTGAGCTGCGAA-MH9143270  IGHV3-30      None   IGHM  IGHJ4      False   \n",
       "27  AACTCAGTCAAAGTAG-MH9143270  IGHV3-23  IGHD3-16   IGHD  IGHJ4      False   \n",
       "30  AACTCCCAGGTGCACA-MH9143270  IGKV6-21      None   IGKC  IGKJ3      False   \n",
       "\n",
       "    full_length                          cdr3  \\\n",
       "3          True                 CGTWDSSLSAWVF   \n",
       "13         True                          None   \n",
       "23         True           CESGRATITMIVVTDLDYW   \n",
       "27         True  CAKAWIGLDSEIYYDYIWGSYRIDFDYW   \n",
       "30         True                          None   \n",
       "\n",
       "                                              cdr3_nt  \n",
       "3             TGCGGAACATGGGATAGCAGCCTGAGTGCTTGGGTGTTC  \n",
       "13                                               None  \n",
       "23  TGCGAAAGCGGGCGGGCTACGATTACTATGATAGTAGTGACTGACC...  \n",
       "27  TGTGCGAAAGCTTGGATCGGACTCGACTCCGAGATTTATTATGATT...  \n",
       "30                                               None  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns += ['cdr3', 'cdr3_nt']\n",
    "df_bcr_raw[(df_bcr_raw['productive']=='False')\n",
    "           & (df_bcr_raw['full_length'])][columns].head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d38cc23f",
   "metadata": {},
   "source": [
    "We will now load the TCR data and convert it from tab-separated format to comma-separated format for later use. Since barcodes overlap between multiple samples within the dataset, we further copy the unique 'CellID' as 'barcode' annotation. This data also contains annotation on patient-level (e.g. Age, Ooutcome, ...) and cell-level annotation (cell type) in addition to the contig annotations previously discussed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c3cd8288",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\felix.drost\\Anaconda3\\envs\\covidVac\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3186: DtypeWarning: Columns (38,39,40) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total measurements:  547630\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Centre</th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>Days_from_onset</th>\n",
       "      <th>time_after_LPS</th>\n",
       "      <th>Worst_Clinical_Status</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>leiden</th>\n",
       "      <th>consensus</th>\n",
       "      <th>initial_clustering</th>\n",
       "      <th>study_id</th>\n",
       "      <th>AgeRange</th>\n",
       "      <th>Age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>AAACCTGAGAAACCTA-MH9179824</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAAACCTA-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>512</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV5-1</td>\n",
       "      <td>None</td>\n",
       "      <td>TRBJ2-3</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>26</td>\n",
       "      <td>CD8</td>\n",
       "      <td>CD8</td>\n",
       "      <td>COVID-014</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>AAACCTGAGAAACCTA-MH9179824</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAAACCTA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>573</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV8-4</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAJ31</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>26</td>\n",
       "      <td>CD8</td>\n",
       "      <td>CD8</td>\n",
       "      <td>COVID-014</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>AAACCTGAGAACAATC-MH9143324</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAACAATC-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>635</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV19</td>\n",
       "      <td>None</td>\n",
       "      <td>TRBJ1-5</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>25</td>\n",
       "      <td>gdT</td>\n",
       "      <td>gdT</td>\n",
       "      <td>COVID-029</td>\n",
       "      <td>(20, 29]</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>AAACCTGAGAACTCGG-MH8919329</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAACTCGG-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>499</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV34</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAJ30</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Death</td>\n",
       "      <td>Death</td>\n",
       "      <td>23</td>\n",
       "      <td>Treg</td>\n",
       "      <td>Treg</td>\n",
       "      <td>COVID-045</td>\n",
       "      <td>(70, 79]</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>AAACCTGAGAACTCGG-MH8919329</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAACTCGG-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>496</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV5-1</td>\n",
       "      <td>None</td>\n",
       "      <td>TRBJ2-5</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Death</td>\n",
       "      <td>Death</td>\n",
       "      <td>23</td>\n",
       "      <td>Treg</td>\n",
       "      <td>Treg</td>\n",
       "      <td>COVID-045</td>\n",
       "      <td>(70, 79]</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Centre                     barcode  is_cell  \\\n",
       "0  Newcastle  AAACCTGAGAAACCTA-MH9179824     True   \n",
       "1  Newcastle  AAACCTGAGAAACCTA-MH9179824     True   \n",
       "2  Newcastle  AAACCTGAGAACAATC-MH9143324     True   \n",
       "3  Newcastle  AAACCTGAGAACTCGG-MH8919329     True   \n",
       "4  Newcastle  AAACCTGAGAACTCGG-MH8919329     True   \n",
       "\n",
       "                     contig_id  high_confidence  length chain   v_gene d_gene  \\\n",
       "0  AAACCTGAGAAACCTA-1_contig_2             True     512   TRB  TRBV5-1   None   \n",
       "1  AAACCTGAGAAACCTA-1_contig_1             True     573   TRA  TRAV8-4   None   \n",
       "2  AAACCTGAGAACAATC-1_contig_1             True     635   TRB   TRBV19   None   \n",
       "3  AAACCTGAGAACTCGG-1_contig_1             True     499   TRA   TRAV34   None   \n",
       "4  AAACCTGAGAACTCGG-1_contig_2             True     496   TRB  TRBV5-1   None   \n",
       "\n",
       "    j_gene  ... Days_from_onset  time_after_LPS  Worst_Clinical_Status  \\\n",
       "0  TRBJ2-3  ...               8             NaN               Moderate   \n",
       "1   TRAJ31  ...               8             NaN               Moderate   \n",
       "2  TRBJ1-5  ...              16             NaN               Moderate   \n",
       "3   TRAJ30  ...               5             NaN                  Death   \n",
       "4  TRBJ2-5  ...               5             NaN                  Death   \n",
       "\n",
       "  Outcome leiden  consensus  initial_clustering   study_id  AgeRange Age  \n",
       "0    Home     26        CD8                 CD8  COVID-014  (50, 59]  54  \n",
       "1    Home     26        CD8                 CD8  COVID-014  (50, 59]  54  \n",
       "2    Home     25        gdT                 gdT  COVID-029  (20, 29]  26  \n",
       "3   Death     23       Treg                Treg  COVID-045  (70, 79]  77  \n",
       "4   Death     23       Treg                Treg  COVID-045  (70, 79]  77  \n",
       "\n",
       "[5 rows x 47 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcr_raw = pd.read_csv(path_tcr_input, sep='\\t')\n",
    "df_tcr_raw['barcode'] = df_tcr_raw.pop('CellID')  \n",
    "df_tcr_raw.to_csv(path_tcr_csv)\n",
    "print('Total measurements: ', len(df_tcr_raw))\n",
    "df_tcr_raw.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d158ca55",
   "metadata": {},
   "source": [
    "We will find that this data already has been processed: All contigs are of full length and represent a productive IR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "acd46004",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True    547630\n",
       "Name: full_length, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcr_raw['full_length'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "680afce9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True    547630\n",
       "Name: productive, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcr_raw['productive'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c01632b2",
   "metadata": {},
   "source": [
    "### 3.2. Cell-Aligned Data\n",
    "In order to perform analysis on cell level, we next need to align IR information previously sorted by contig to a cell based format via the cell barcode. Note that usually many contigs map to the same cell, since every IR consists of VJ and VDJ chain. Also it has been shown, that a cell can express multiple IRs [cite].\n",
    "\n",
    "\n",
    "We will utilize the python package scirpy, that performs the cell-allignment automatically, when reading various TCR information from various formats:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "58b88cb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\felix.drost\\Anaconda3\\envs\\covidVac\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3457: DtypeWarning: Columns (38,39,40) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Non-standard locus name ignored: Multi \n",
      "Amount cells:  280045\n"
     ]
    }
   ],
   "source": [
    "adata_tcr = ir.io.read_10x_vdj(path_tcr_csv)\n",
    "print('Amount cells: ', len(adata_tcr))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ba4e170",
   "metadata": {},
   "source": [
    "The IR information will by stored sorted in adata_tcr.obs as a DataFrame with the index representing an individual cell. The DataFrame contains the following columns:\n",
    "- has_ir: indicates whether an IR was detected\n",
    "- multi_chain: indicates whether more than 2 IRs were detected\n",
    "- extra_chains: list of additional chains if multi_chain is true\n",
    "\n",
    "The contig data we observed in the raw data is stored as well. It is stored as seperate columns for the two chains (VJ and VDJ) with up to 2 contigs per chain for each cell:\n",
    "- IR_V{D}J_{1,2}_locus\n",
    "- IR_V{D}J_{1,2}_productive\n",
    "- IR_V{D}J_{1,2}_{v,d,j,c}_call\n",
    "- IR_V{D}J_{1,2}_junction{_aa}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "22313f08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>is_cell</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>multi_chain</th>\n",
       "      <th>extra_chains</th>\n",
       "      <th>IR_VJ_1_c_call</th>\n",
       "      <th>IR_VJ_2_c_call</th>\n",
       "      <th>IR_VDJ_1_c_call</th>\n",
       "      <th>IR_VDJ_2_c_call</th>\n",
       "      <th>IR_VJ_1_consensus_count</th>\n",
       "      <th>IR_VJ_2_consensus_count</th>\n",
       "      <th>...</th>\n",
       "      <th>IR_VDJ_2_locus</th>\n",
       "      <th>IR_VJ_1_productive</th>\n",
       "      <th>IR_VJ_2_productive</th>\n",
       "      <th>IR_VDJ_1_productive</th>\n",
       "      <th>IR_VDJ_2_productive</th>\n",
       "      <th>IR_VJ_1_v_call</th>\n",
       "      <th>IR_VJ_2_v_call</th>\n",
       "      <th>IR_VDJ_1_v_call</th>\n",
       "      <th>IR_VDJ_2_v_call</th>\n",
       "      <th>has_ir</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACCTA-MH9179824</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>960.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAV8-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV5-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACAATC-MH9143324</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACTCGG-MH8919329</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5122.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAV34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV5-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGCCCA-MH9143373</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25914.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAV16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGACA-MH9143274</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3212.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAV8-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV7-6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           is_cell high_confidence multi_chain extra_chains  \\\n",
       "cell_id                                                                       \n",
       "AAACCTGAGAAACCTA-MH9179824    True            True       False           []   \n",
       "AAACCTGAGAACAATC-MH9143324    True            True       False           []   \n",
       "AAACCTGAGAACTCGG-MH8919329    True            True       False           []   \n",
       "AAACCTGAGAAGCCCA-MH9143373    True            True       False           []   \n",
       "AAACCTGAGAAGGACA-MH9143274    True            True       False           []   \n",
       "\n",
       "                           IR_VJ_1_c_call IR_VJ_2_c_call IR_VDJ_1_c_call  \\\n",
       "cell_id                                                                    \n",
       "AAACCTGAGAAACCTA-MH9179824           TRAC            NaN           TRBC2   \n",
       "AAACCTGAGAACAATC-MH9143324            NaN            NaN           TRBC1   \n",
       "AAACCTGAGAACTCGG-MH8919329           TRAC            NaN           TRBC2   \n",
       "AAACCTGAGAAGCCCA-MH9143373           TRAC            NaN             NaN   \n",
       "AAACCTGAGAAGGACA-MH9143274           TRAC            NaN           TRBC2   \n",
       "\n",
       "                           IR_VDJ_2_c_call  IR_VJ_1_consensus_count  \\\n",
       "cell_id                                                               \n",
       "AAACCTGAGAAACCTA-MH9179824             NaN                    960.0   \n",
       "AAACCTGAGAACAATC-MH9143324             NaN                      NaN   \n",
       "AAACCTGAGAACTCGG-MH8919329             NaN                   5122.0   \n",
       "AAACCTGAGAAGCCCA-MH9143373             NaN                  25914.0   \n",
       "AAACCTGAGAAGGACA-MH9143274             NaN                   3212.0   \n",
       "\n",
       "                            IR_VJ_2_consensus_count  ...  IR_VDJ_2_locus  \\\n",
       "cell_id                                              ...                   \n",
       "AAACCTGAGAAACCTA-MH9179824                      NaN  ...             NaN   \n",
       "AAACCTGAGAACAATC-MH9143324                      NaN  ...             NaN   \n",
       "AAACCTGAGAACTCGG-MH8919329                      NaN  ...             NaN   \n",
       "AAACCTGAGAAGCCCA-MH9143373                      NaN  ...             NaN   \n",
       "AAACCTGAGAAGGACA-MH9143274                      NaN  ...             NaN   \n",
       "\n",
       "                            IR_VJ_1_productive IR_VJ_2_productive  \\\n",
       "cell_id                                                             \n",
       "AAACCTGAGAAACCTA-MH9179824                True               None   \n",
       "AAACCTGAGAACAATC-MH9143324                None               None   \n",
       "AAACCTGAGAACTCGG-MH8919329                True               None   \n",
       "AAACCTGAGAAGCCCA-MH9143373                True               None   \n",
       "AAACCTGAGAAGGACA-MH9143274                True               None   \n",
       "\n",
       "                           IR_VDJ_1_productive IR_VDJ_2_productive  \\\n",
       "cell_id                                                              \n",
       "AAACCTGAGAAACCTA-MH9179824                True                None   \n",
       "AAACCTGAGAACAATC-MH9143324                True                None   \n",
       "AAACCTGAGAACTCGG-MH8919329                True                None   \n",
       "AAACCTGAGAAGCCCA-MH9143373                None                None   \n",
       "AAACCTGAGAAGGACA-MH9143274                True                None   \n",
       "\n",
       "                           IR_VJ_1_v_call  IR_VJ_2_v_call  IR_VDJ_1_v_call  \\\n",
       "cell_id                                                                      \n",
       "AAACCTGAGAAACCTA-MH9179824        TRAV8-4             NaN          TRBV5-1   \n",
       "AAACCTGAGAACAATC-MH9143324            NaN             NaN           TRBV19   \n",
       "AAACCTGAGAACTCGG-MH8919329         TRAV34             NaN          TRBV5-1   \n",
       "AAACCTGAGAAGCCCA-MH9143373         TRAV16             NaN              NaN   \n",
       "AAACCTGAGAAGGACA-MH9143274        TRAV8-1             NaN          TRBV7-6   \n",
       "\n",
       "                            IR_VDJ_2_v_call  has_ir  \n",
       "cell_id                                              \n",
       "AAACCTGAGAAACCTA-MH9179824              NaN    True  \n",
       "AAACCTGAGAACAATC-MH9143324              NaN    True  \n",
       "AAACCTGAGAACTCGG-MH8919329              NaN    True  \n",
       "AAACCTGAGAAGCCCA-MH9143373              NaN    True  \n",
       "AAACCTGAGAAGGACA-MH9143274              NaN    True  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_tcr.obs.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bd32e8b",
   "metadata": {},
   "source": [
    "Notice, that the patient-level information is not automatically added here. Let's add them by loading the raw data, alligning them on a cell level and indexing them by their barcode."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1c81b023",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Centre</th>\n",
       "      <th>Sample</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>Collection_Day</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Swab_result</th>\n",
       "      <th>Status</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Status_on_day_collection</th>\n",
       "      <th>Status_on_day_collection_summary</th>\n",
       "      <th>Days_from_onset</th>\n",
       "      <th>time_after_LPS</th>\n",
       "      <th>Worst_Clinical_Status</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>initial_clustering</th>\n",
       "      <th>study_id</th>\n",
       "      <th>AgeRange</th>\n",
       "      <th>Age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACCTA-MH9179824</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>MH9179824</td>\n",
       "      <td>COVID-014</td>\n",
       "      <td>D0</td>\n",
       "      <td>Female</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Covid</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD8</td>\n",
       "      <td>COVID-014</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACAATC-MH9143324</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>MH9143324</td>\n",
       "      <td>COVID-029</td>\n",
       "      <td>D0</td>\n",
       "      <td>Male</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Covid</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>gdT</td>\n",
       "      <td>COVID-029</td>\n",
       "      <td>(20, 29]</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACTCGG-MH8919329</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>MH8919329</td>\n",
       "      <td>COVID-045</td>\n",
       "      <td>D0</td>\n",
       "      <td>Female</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Covid</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>ITU_intubated</td>\n",
       "      <td>Critical</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Death</td>\n",
       "      <td>Death</td>\n",
       "      <td>Treg</td>\n",
       "      <td>COVID-045</td>\n",
       "      <td>(70, 79]</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGCCCA-MH9143373</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>MH9143373</td>\n",
       "      <td>COVID-060</td>\n",
       "      <td>D0</td>\n",
       "      <td>Male</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Non_covid</td>\n",
       "      <td>Smoker</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Non_covid</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non-covid</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD4</td>\n",
       "      <td>COVID-060</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGACA-MH9143274</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>MH9143274</td>\n",
       "      <td>COVID-037</td>\n",
       "      <td>D0</td>\n",
       "      <td>Female</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Covid</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>ITU_O2</td>\n",
       "      <td>Severe</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD4</td>\n",
       "      <td>COVID-037</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Centre     Sample patient_id Collection_Day  \\\n",
       "AAACCTGAGAAACCTA-MH9179824  Newcastle  MH9179824  COVID-014             D0   \n",
       "AAACCTGAGAACAATC-MH9143324  Newcastle  MH9143324  COVID-029             D0   \n",
       "AAACCTGAGAACTCGG-MH8919329  Newcastle  MH8919329  COVID-045             D0   \n",
       "AAACCTGAGAAGCCCA-MH9143373  Newcastle  MH9143373  COVID-060             D0   \n",
       "AAACCTGAGAAGGACA-MH9143274  Newcastle  MH9143274  COVID-037             D0   \n",
       "\n",
       "                               Sex Swab_result     Status      Smoker  \\\n",
       "AAACCTGAGAAACCTA-MH9179824  Female    Positive      Covid  Non-smoker   \n",
       "AAACCTGAGAACAATC-MH9143324    Male    Positive      Covid  Non-smoker   \n",
       "AAACCTGAGAACTCGG-MH8919329  Female    Positive      Covid  Non-smoker   \n",
       "AAACCTGAGAAGCCCA-MH9143373    Male    Negative  Non_covid      Smoker   \n",
       "AAACCTGAGAAGGACA-MH9143274  Female    Positive      Covid  Non-smoker   \n",
       "\n",
       "                           Status_on_day_collection  \\\n",
       "AAACCTGAGAAACCTA-MH9179824                  Ward_O2   \n",
       "AAACCTGAGAACAATC-MH9143324                  Ward_O2   \n",
       "AAACCTGAGAACTCGG-MH8919329            ITU_intubated   \n",
       "AAACCTGAGAAGCCCA-MH9143373                  Ward_O2   \n",
       "AAACCTGAGAAGGACA-MH9143274                   ITU_O2   \n",
       "\n",
       "                           Status_on_day_collection_summary Days_from_onset  \\\n",
       "AAACCTGAGAAACCTA-MH9179824                         Moderate               8   \n",
       "AAACCTGAGAACAATC-MH9143324                         Moderate              16   \n",
       "AAACCTGAGAACTCGG-MH8919329                         Critical               5   \n",
       "AAACCTGAGAAGCCCA-MH9143373                        Non_covid               3   \n",
       "AAACCTGAGAAGGACA-MH9143274                           Severe              21   \n",
       "\n",
       "                           time_after_LPS Worst_Clinical_Status Outcome  \\\n",
       "AAACCTGAGAAACCTA-MH9179824            NaN              Moderate    Home   \n",
       "AAACCTGAGAACAATC-MH9143324            NaN              Moderate    Home   \n",
       "AAACCTGAGAACTCGG-MH8919329            NaN                 Death   Death   \n",
       "AAACCTGAGAAGCCCA-MH9143373            NaN             Non-covid    Home   \n",
       "AAACCTGAGAAGGACA-MH9143274            NaN                Severe    Home   \n",
       "\n",
       "                           initial_clustering   study_id  AgeRange  Age  \n",
       "AAACCTGAGAAACCTA-MH9179824                CD8  COVID-014  (50, 59]   54  \n",
       "AAACCTGAGAACAATC-MH9143324                gdT  COVID-029  (20, 29]   26  \n",
       "AAACCTGAGAACTCGG-MH8919329               Treg  COVID-045  (70, 79]   77  \n",
       "AAACCTGAGAAGCCCA-MH9143373                CD4  COVID-060  (50, 59]   54  \n",
       "AAACCTGAGAAGGACA-MH9143274                CD4  COVID-037  (50, 59]   59  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_information = [\n",
    "    'barcode', 'Centre', 'Sample', 'patient_id', 'Collection_Day', 'Sex', \n",
    "    'Swab_result', 'Status', 'Smoker', 'Status_on_day_collection',\n",
    "    'Status_on_day_collection_summary', 'Days_from_onset', 'time_after_LPS',\n",
    "    'Worst_Clinical_Status', 'Outcome',\n",
    "    'initial_clustering', 'study_id', 'AgeRange', 'Age']\n",
    "df_patient = df_tcr_raw[patient_information].copy()\n",
    "df_patient['Days_from_onset'] = df_patient['Days_from_onset'].astype(str)  # mixed type (str, int)\n",
    "df_patient = df_patient.drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Assigning barcode as index\n",
    "df_patient.index = df_patient.pop('barcode')\n",
    "df_patient.index.name = None\n",
    "df_patient.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "577fe80b",
   "metadata": {},
   "source": [
    "We can now add this information to the adata object and check, whether the annotation is added as observation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e43ff4fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>is_cell</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>multi_chain</th>\n",
       "      <th>extra_chains</th>\n",
       "      <th>IR_VJ_1_c_call</th>\n",
       "      <th>IR_VJ_2_c_call</th>\n",
       "      <th>IR_VDJ_1_c_call</th>\n",
       "      <th>IR_VDJ_2_c_call</th>\n",
       "      <th>IR_VJ_1_consensus_count</th>\n",
       "      <th>IR_VJ_2_consensus_count</th>\n",
       "      <th>...</th>\n",
       "      <th>Status_on_day_collection</th>\n",
       "      <th>Status_on_day_collection_summary</th>\n",
       "      <th>Days_from_onset</th>\n",
       "      <th>time_after_LPS</th>\n",
       "      <th>Worst_Clinical_Status</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>initial_clustering</th>\n",
       "      <th>study_id</th>\n",
       "      <th>AgeRange</th>\n",
       "      <th>Age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACCTA-MH9179824</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>960.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD8</td>\n",
       "      <td>COVID-014</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACAATC-MH9143324</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>gdT</td>\n",
       "      <td>COVID-029</td>\n",
       "      <td>(20, 29]</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACTCGG-MH8919329</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5122.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ITU_intubated</td>\n",
       "      <td>Critical</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Death</td>\n",
       "      <td>Death</td>\n",
       "      <td>Treg</td>\n",
       "      <td>COVID-045</td>\n",
       "      <td>(70, 79]</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGCCCA-MH9143373</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25914.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Non_covid</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non-covid</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD4</td>\n",
       "      <td>COVID-060</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGACA-MH9143274</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3212.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ITU_O2</td>\n",
       "      <td>Severe</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD4</td>\n",
       "      <td>COVID-037</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           is_cell high_confidence multi_chain extra_chains  \\\n",
       "cell_id                                                                       \n",
       "AAACCTGAGAAACCTA-MH9179824    True            True       False           []   \n",
       "AAACCTGAGAACAATC-MH9143324    True            True       False           []   \n",
       "AAACCTGAGAACTCGG-MH8919329    True            True       False           []   \n",
       "AAACCTGAGAAGCCCA-MH9143373    True            True       False           []   \n",
       "AAACCTGAGAAGGACA-MH9143274    True            True       False           []   \n",
       "\n",
       "                           IR_VJ_1_c_call IR_VJ_2_c_call IR_VDJ_1_c_call  \\\n",
       "cell_id                                                                    \n",
       "AAACCTGAGAAACCTA-MH9179824           TRAC            NaN           TRBC2   \n",
       "AAACCTGAGAACAATC-MH9143324            NaN            NaN           TRBC1   \n",
       "AAACCTGAGAACTCGG-MH8919329           TRAC            NaN           TRBC2   \n",
       "AAACCTGAGAAGCCCA-MH9143373           TRAC            NaN             NaN   \n",
       "AAACCTGAGAAGGACA-MH9143274           TRAC            NaN           TRBC2   \n",
       "\n",
       "                           IR_VDJ_2_c_call  IR_VJ_1_consensus_count  \\\n",
       "cell_id                                                               \n",
       "AAACCTGAGAAACCTA-MH9179824             NaN                    960.0   \n",
       "AAACCTGAGAACAATC-MH9143324             NaN                      NaN   \n",
       "AAACCTGAGAACTCGG-MH8919329             NaN                   5122.0   \n",
       "AAACCTGAGAAGCCCA-MH9143373             NaN                  25914.0   \n",
       "AAACCTGAGAAGGACA-MH9143274             NaN                   3212.0   \n",
       "\n",
       "                            IR_VJ_2_consensus_count  ...  \\\n",
       "cell_id                                              ...   \n",
       "AAACCTGAGAAACCTA-MH9179824                      NaN  ...   \n",
       "AAACCTGAGAACAATC-MH9143324                      NaN  ...   \n",
       "AAACCTGAGAACTCGG-MH8919329                      NaN  ...   \n",
       "AAACCTGAGAAGCCCA-MH9143373                      NaN  ...   \n",
       "AAACCTGAGAAGGACA-MH9143274                      NaN  ...   \n",
       "\n",
       "                            Status_on_day_collection  \\\n",
       "cell_id                                                \n",
       "AAACCTGAGAAACCTA-MH9179824                   Ward_O2   \n",
       "AAACCTGAGAACAATC-MH9143324                   Ward_O2   \n",
       "AAACCTGAGAACTCGG-MH8919329             ITU_intubated   \n",
       "AAACCTGAGAAGCCCA-MH9143373                   Ward_O2   \n",
       "AAACCTGAGAAGGACA-MH9143274                    ITU_O2   \n",
       "\n",
       "                            Status_on_day_collection_summary Days_from_onset  \\\n",
       "cell_id                                                                        \n",
       "AAACCTGAGAAACCTA-MH9179824                          Moderate               8   \n",
       "AAACCTGAGAACAATC-MH9143324                          Moderate              16   \n",
       "AAACCTGAGAACTCGG-MH8919329                          Critical               5   \n",
       "AAACCTGAGAAGCCCA-MH9143373                         Non_covid               3   \n",
       "AAACCTGAGAAGGACA-MH9143274                            Severe              21   \n",
       "\n",
       "                           time_after_LPS Worst_Clinical_Status Outcome  \\\n",
       "cell_id                                                                   \n",
       "AAACCTGAGAAACCTA-MH9179824            NaN              Moderate    Home   \n",
       "AAACCTGAGAACAATC-MH9143324            NaN              Moderate    Home   \n",
       "AAACCTGAGAACTCGG-MH8919329            NaN                 Death   Death   \n",
       "AAACCTGAGAAGCCCA-MH9143373            NaN             Non-covid    Home   \n",
       "AAACCTGAGAAGGACA-MH9143274            NaN                Severe    Home   \n",
       "\n",
       "                            initial_clustering   study_id  AgeRange  Age  \n",
       "cell_id                                                                   \n",
       "AAACCTGAGAAACCTA-MH9179824                 CD8  COVID-014  (50, 59]   54  \n",
       "AAACCTGAGAACAATC-MH9143324                 gdT  COVID-029  (20, 29]   26  \n",
       "AAACCTGAGAACTCGG-MH8919329                Treg  COVID-045  (70, 79]   77  \n",
       "AAACCTGAGAAGCCCA-MH9143373                 CD4  COVID-060  (50, 59]   54  \n",
       "AAACCTGAGAAGGACA-MH9143274                 CD4  COVID-037  (50, 59]   59  \n",
       "\n",
       "[5 rows x 63 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_tcr.obs[df_patient.columns] = df_patient\n",
    "adata_tcr.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f62989d4",
   "metadata": {},
   "source": [
    "In the same manner we can load the BCR data and check the IR annotation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "248e0fcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\felix.drost\\Anaconda3\\envs\\covidVac\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3457: DtypeWarning: Columns (12,20) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Non-standard locus name ignored: Multi \n",
      "WARNING: Non-standard locus name ignored: None \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>is_cell</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>multi_chain</th>\n",
       "      <th>extra_chains</th>\n",
       "      <th>IR_VJ_1_c_call</th>\n",
       "      <th>IR_VJ_2_c_call</th>\n",
       "      <th>IR_VDJ_1_c_call</th>\n",
       "      <th>IR_VDJ_2_c_call</th>\n",
       "      <th>IR_VJ_1_consensus_count</th>\n",
       "      <th>IR_VJ_2_consensus_count</th>\n",
       "      <th>...</th>\n",
       "      <th>IR_VDJ_2_locus</th>\n",
       "      <th>IR_VJ_1_productive</th>\n",
       "      <th>IR_VJ_2_productive</th>\n",
       "      <th>IR_VDJ_1_productive</th>\n",
       "      <th>IR_VDJ_2_productive</th>\n",
       "      <th>IR_VJ_1_v_call</th>\n",
       "      <th>IR_VJ_2_v_call</th>\n",
       "      <th>IR_VDJ_1_v_call</th>\n",
       "      <th>IR_VDJ_2_v_call</th>\n",
       "      <th>has_ir</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACTCGG-MH9143420</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6982.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKV3-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATAGGG-MH8919227</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>626.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKV3-15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHV1-24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATGTGT-MH8919326</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[{\"c_call\": \"IGLC1\", \"consensus_count\": 748, \"...</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1028.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKV1-33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHV1-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATGTGT-MH9143424</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>IGLC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>736.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLV3-21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGACACTAA-MH9143423</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>66.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKV3-20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           is_cell high_confidence multi_chain  \\\n",
       "cell_id                                                          \n",
       "AAACCTGAGAACTCGG-MH9143420    True            True       False   \n",
       "AAACCTGAGAATAGGG-MH8919227    True            True       False   \n",
       "AAACCTGAGAATGTGT-MH8919326    True            True       False   \n",
       "AAACCTGAGAATGTGT-MH9143424    True            True       False   \n",
       "AAACCTGAGACACTAA-MH9143423    True            True       False   \n",
       "\n",
       "                                                                 extra_chains  \\\n",
       "cell_id                                                                         \n",
       "AAACCTGAGAACTCGG-MH9143420                                                 []   \n",
       "AAACCTGAGAATAGGG-MH8919227                                                 []   \n",
       "AAACCTGAGAATGTGT-MH8919326  [{\"c_call\": \"IGLC1\", \"consensus_count\": 748, \"...   \n",
       "AAACCTGAGAATGTGT-MH9143424                                                 []   \n",
       "AAACCTGAGACACTAA-MH9143423                                                 []   \n",
       "\n",
       "                           IR_VJ_1_c_call IR_VJ_2_c_call IR_VDJ_1_c_call  \\\n",
       "cell_id                                                                    \n",
       "AAACCTGAGAACTCGG-MH9143420           IGKC            NaN            IGHM   \n",
       "AAACCTGAGAATAGGG-MH8919227           IGKC            NaN           IGHA1   \n",
       "AAACCTGAGAATGTGT-MH8919326           IGKC            NaN            IGHM   \n",
       "AAACCTGAGAATGTGT-MH9143424          IGLC2            NaN             NaN   \n",
       "AAACCTGAGACACTAA-MH9143423           IGKC            NaN           IGHG1   \n",
       "\n",
       "                           IR_VDJ_2_c_call  IR_VJ_1_consensus_count  \\\n",
       "cell_id                                                               \n",
       "AAACCTGAGAACTCGG-MH9143420             NaN                   6982.0   \n",
       "AAACCTGAGAATAGGG-MH8919227             NaN                    626.0   \n",
       "AAACCTGAGAATGTGT-MH8919326             NaN                   1028.0   \n",
       "AAACCTGAGAATGTGT-MH9143424             NaN                    736.0   \n",
       "AAACCTGAGACACTAA-MH9143423             NaN                     66.0   \n",
       "\n",
       "                            IR_VJ_2_consensus_count  ...  IR_VDJ_2_locus  \\\n",
       "cell_id                                              ...                   \n",
       "AAACCTGAGAACTCGG-MH9143420                      NaN  ...             NaN   \n",
       "AAACCTGAGAATAGGG-MH8919227                      NaN  ...             NaN   \n",
       "AAACCTGAGAATGTGT-MH8919326                      NaN  ...             NaN   \n",
       "AAACCTGAGAATGTGT-MH9143424                      NaN  ...             NaN   \n",
       "AAACCTGAGACACTAA-MH9143423                      NaN  ...             NaN   \n",
       "\n",
       "                            IR_VJ_1_productive IR_VJ_2_productive  \\\n",
       "cell_id                                                             \n",
       "AAACCTGAGAACTCGG-MH9143420                True               None   \n",
       "AAACCTGAGAATAGGG-MH8919227                True               None   \n",
       "AAACCTGAGAATGTGT-MH8919326                True               None   \n",
       "AAACCTGAGAATGTGT-MH9143424                True               None   \n",
       "AAACCTGAGACACTAA-MH9143423                True               None   \n",
       "\n",
       "                           IR_VDJ_1_productive IR_VDJ_2_productive  \\\n",
       "cell_id                                                              \n",
       "AAACCTGAGAACTCGG-MH9143420                True                None   \n",
       "AAACCTGAGAATAGGG-MH8919227                True                None   \n",
       "AAACCTGAGAATGTGT-MH8919326                True                None   \n",
       "AAACCTGAGAATGTGT-MH9143424                None                None   \n",
       "AAACCTGAGACACTAA-MH9143423                True                None   \n",
       "\n",
       "                           IR_VJ_1_v_call  IR_VJ_2_v_call  IR_VDJ_1_v_call  \\\n",
       "cell_id                                                                      \n",
       "AAACCTGAGAACTCGG-MH9143420       IGKV3-11             NaN         IGHV3-23   \n",
       "AAACCTGAGAATAGGG-MH8919227       IGKV3-15             NaN         IGHV1-24   \n",
       "AAACCTGAGAATGTGT-MH8919326       IGKV1-33             NaN          IGHV1-2   \n",
       "AAACCTGAGAATGTGT-MH9143424       IGLV3-21             NaN              NaN   \n",
       "AAACCTGAGACACTAA-MH9143423       IGKV3-20             NaN         IGHV3-23   \n",
       "\n",
       "                            IR_VDJ_2_v_call  has_ir  \n",
       "cell_id                                              \n",
       "AAACCTGAGAACTCGG-MH9143420              NaN    True  \n",
       "AAACCTGAGAATAGGG-MH8919227              NaN    True  \n",
       "AAACCTGAGAATGTGT-MH8919326              NaN    True  \n",
       "AAACCTGAGAATGTGT-MH9143424              NaN    True  \n",
       "AAACCTGAGACACTAA-MH9143423              NaN    True  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_bcr = ir.io.read_10x_vdj(path_bcr_input)\n",
    "adata_bcr.obs.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2411023c",
   "metadata": {},
   "source": [
    "As before, we can add extract additional information annotated for each contig."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8c4e5c02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCGATATA-MH9143270</th>\n",
       "      <td>COVID-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCTACCAGA-MH9143270</th>\n",
       "      <td>COVID-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGCATTTGCTT-MH9143270</th>\n",
       "      <td>COVID-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGGTAGCACGA-MH9143270</th>\n",
       "      <td>COVID-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGTCGTACCGG-MH9143270</th>\n",
       "      <td>COVID-002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           patient_id\n",
       "AAACCTGAGCGATATA-MH9143270  COVID-002\n",
       "AAACCTGTCTACCAGA-MH9143270  COVID-002\n",
       "AAACGGGCATTTGCTT-MH9143270  COVID-002\n",
       "AAACGGGGTAGCACGA-MH9143270  COVID-002\n",
       "AAACGGGTCGTACCGG-MH9143270  COVID-002"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_information = ['barcode', 'patient_id']\n",
    "df_patient = df_bcr_raw[patient_information].copy()\n",
    "df_patient = df_patient.drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Assigning barcode as index\n",
    "df_patient.index = df_patient.pop('barcode')\n",
    "df_patient.index.name = None\n",
    "df_patient.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4038ae1",
   "metadata": {},
   "source": [
    "And add them to the cell annotations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ee57129a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>is_cell</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>multi_chain</th>\n",
       "      <th>extra_chains</th>\n",
       "      <th>IR_VJ_1_c_call</th>\n",
       "      <th>IR_VJ_2_c_call</th>\n",
       "      <th>IR_VDJ_1_c_call</th>\n",
       "      <th>IR_VDJ_2_c_call</th>\n",
       "      <th>IR_VJ_1_consensus_count</th>\n",
       "      <th>IR_VJ_2_consensus_count</th>\n",
       "      <th>...</th>\n",
       "      <th>IR_VJ_1_productive</th>\n",
       "      <th>IR_VJ_2_productive</th>\n",
       "      <th>IR_VDJ_1_productive</th>\n",
       "      <th>IR_VDJ_2_productive</th>\n",
       "      <th>IR_VJ_1_v_call</th>\n",
       "      <th>IR_VJ_2_v_call</th>\n",
       "      <th>IR_VDJ_1_v_call</th>\n",
       "      <th>IR_VDJ_2_v_call</th>\n",
       "      <th>has_ir</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACTCGG-MH9143420</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6982.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKV3-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>COVID-011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATAGGG-MH8919227</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>626.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKV3-15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHV1-24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>COVID-075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATGTGT-MH8919326</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[{\"c_call\": \"IGLC1\", \"consensus_count\": 748, \"...</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1028.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKV1-33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHV1-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>COVID-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATGTGT-MH9143424</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>IGLC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>736.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLV3-21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>COVID-039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGACACTAA-MH9143423</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>66.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKV3-20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>COVID-036</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           is_cell high_confidence multi_chain  \\\n",
       "cell_id                                                          \n",
       "AAACCTGAGAACTCGG-MH9143420    True            True       False   \n",
       "AAACCTGAGAATAGGG-MH8919227    True            True       False   \n",
       "AAACCTGAGAATGTGT-MH8919326    True            True       False   \n",
       "AAACCTGAGAATGTGT-MH9143424    True            True       False   \n",
       "AAACCTGAGACACTAA-MH9143423    True            True       False   \n",
       "\n",
       "                                                                 extra_chains  \\\n",
       "cell_id                                                                         \n",
       "AAACCTGAGAACTCGG-MH9143420                                                 []   \n",
       "AAACCTGAGAATAGGG-MH8919227                                                 []   \n",
       "AAACCTGAGAATGTGT-MH8919326  [{\"c_call\": \"IGLC1\", \"consensus_count\": 748, \"...   \n",
       "AAACCTGAGAATGTGT-MH9143424                                                 []   \n",
       "AAACCTGAGACACTAA-MH9143423                                                 []   \n",
       "\n",
       "                           IR_VJ_1_c_call IR_VJ_2_c_call IR_VDJ_1_c_call  \\\n",
       "cell_id                                                                    \n",
       "AAACCTGAGAACTCGG-MH9143420           IGKC            NaN            IGHM   \n",
       "AAACCTGAGAATAGGG-MH8919227           IGKC            NaN           IGHA1   \n",
       "AAACCTGAGAATGTGT-MH8919326           IGKC            NaN            IGHM   \n",
       "AAACCTGAGAATGTGT-MH9143424          IGLC2            NaN             NaN   \n",
       "AAACCTGAGACACTAA-MH9143423           IGKC            NaN           IGHG1   \n",
       "\n",
       "                           IR_VDJ_2_c_call  IR_VJ_1_consensus_count  \\\n",
       "cell_id                                                               \n",
       "AAACCTGAGAACTCGG-MH9143420             NaN                   6982.0   \n",
       "AAACCTGAGAATAGGG-MH8919227             NaN                    626.0   \n",
       "AAACCTGAGAATGTGT-MH8919326             NaN                   1028.0   \n",
       "AAACCTGAGAATGTGT-MH9143424             NaN                    736.0   \n",
       "AAACCTGAGACACTAA-MH9143423             NaN                     66.0   \n",
       "\n",
       "                            IR_VJ_2_consensus_count  ...  IR_VJ_1_productive  \\\n",
       "cell_id                                              ...                       \n",
       "AAACCTGAGAACTCGG-MH9143420                      NaN  ...                True   \n",
       "AAACCTGAGAATAGGG-MH8919227                      NaN  ...                True   \n",
       "AAACCTGAGAATGTGT-MH8919326                      NaN  ...                True   \n",
       "AAACCTGAGAATGTGT-MH9143424                      NaN  ...                True   \n",
       "AAACCTGAGACACTAA-MH9143423                      NaN  ...                True   \n",
       "\n",
       "                            IR_VJ_2_productive IR_VDJ_1_productive  \\\n",
       "cell_id                                                              \n",
       "AAACCTGAGAACTCGG-MH9143420                None                True   \n",
       "AAACCTGAGAATAGGG-MH8919227                None                True   \n",
       "AAACCTGAGAATGTGT-MH8919326                None                True   \n",
       "AAACCTGAGAATGTGT-MH9143424                None                None   \n",
       "AAACCTGAGACACTAA-MH9143423                None                True   \n",
       "\n",
       "                           IR_VDJ_2_productive IR_VJ_1_v_call IR_VJ_2_v_call  \\\n",
       "cell_id                                                                        \n",
       "AAACCTGAGAACTCGG-MH9143420                None       IGKV3-11            NaN   \n",
       "AAACCTGAGAATAGGG-MH8919227                None       IGKV3-15            NaN   \n",
       "AAACCTGAGAATGTGT-MH8919326                None       IGKV1-33            NaN   \n",
       "AAACCTGAGAATGTGT-MH9143424                None       IGLV3-21            NaN   \n",
       "AAACCTGAGACACTAA-MH9143423                None       IGKV3-20            NaN   \n",
       "\n",
       "                            IR_VDJ_1_v_call  IR_VDJ_2_v_call  has_ir  \\\n",
       "cell_id                                                                \n",
       "AAACCTGAGAACTCGG-MH9143420         IGHV3-23              NaN    True   \n",
       "AAACCTGAGAATAGGG-MH8919227         IGHV1-24              NaN    True   \n",
       "AAACCTGAGAATGTGT-MH8919326          IGHV1-2              NaN    True   \n",
       "AAACCTGAGAATGTGT-MH9143424              NaN              NaN    True   \n",
       "AAACCTGAGACACTAA-MH9143423         IGHV3-23              NaN    True   \n",
       "\n",
       "                            patient_id  \n",
       "cell_id                                 \n",
       "AAACCTGAGAACTCGG-MH9143420   COVID-011  \n",
       "AAACCTGAGAATAGGG-MH8919227   COVID-075  \n",
       "AAACCTGAGAATGTGT-MH8919326   COVID-003  \n",
       "AAACCTGAGAATGTGT-MH9143424   COVID-039  \n",
       "AAACCTGAGACACTAA-MH9143423   COVID-036  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_bcr.obs[df_patient.columns] = df_patient\n",
    "adata_bcr.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bfad98f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 159446 × 0\n",
       "    obs: 'is_cell', 'high_confidence', 'multi_chain', 'extra_chains', 'IR_VJ_1_c_call', 'IR_VJ_2_c_call', 'IR_VDJ_1_c_call', 'IR_VDJ_2_c_call', 'IR_VJ_1_consensus_count', 'IR_VJ_2_consensus_count', 'IR_VDJ_1_consensus_count', 'IR_VDJ_2_consensus_count', 'IR_VJ_1_d_call', 'IR_VJ_2_d_call', 'IR_VDJ_1_d_call', 'IR_VDJ_2_d_call', 'IR_VJ_1_duplicate_count', 'IR_VJ_2_duplicate_count', 'IR_VDJ_1_duplicate_count', 'IR_VDJ_2_duplicate_count', 'IR_VJ_1_j_call', 'IR_VJ_2_j_call', 'IR_VDJ_1_j_call', 'IR_VDJ_2_j_call', 'IR_VJ_1_junction', 'IR_VJ_2_junction', 'IR_VDJ_1_junction', 'IR_VDJ_2_junction', 'IR_VJ_1_junction_aa', 'IR_VJ_2_junction_aa', 'IR_VDJ_1_junction_aa', 'IR_VDJ_2_junction_aa', 'IR_VJ_1_locus', 'IR_VJ_2_locus', 'IR_VDJ_1_locus', 'IR_VDJ_2_locus', 'IR_VJ_1_productive', 'IR_VJ_2_productive', 'IR_VDJ_1_productive', 'IR_VDJ_2_productive', 'IR_VJ_1_v_call', 'IR_VJ_2_v_call', 'IR_VDJ_1_v_call', 'IR_VDJ_2_v_call', 'has_ir', 'patient_id'\n",
       "    uns: 'scirpy_version'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_bcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5a543512",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO discuss downsampling for faster runs?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b26c6edb",
   "metadata": {},
   "source": [
    "## 4. Assinging IR State"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ab09176",
   "metadata": {},
   "source": [
    "In this allignment it can happen that a cell does not get matched a VJ and VDJ chain. On the other hand it is also possible that a cell gets assigned multiple tcrs. While it has been observed that T / B cells can express dual IR, cells with >2 IRs can be indicative as dublets. \n",
    "\n",
    "In order to check quality: Assign IR state"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "af4c993b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\felix.drost\\Anaconda3\\envs\\covidVac\\lib\\site-packages\\ipykernel_launcher.py:1: FutureWarning: chain_pairing is a deprecated function and will be removed in a future version of scirpy. Use `tl.chain_qc` instead.\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "single pair        196957\n",
       "orphan VDJ          45266\n",
       "extra VJ            19034\n",
       "orphan VJ            7937\n",
       "extra VDJ            7473\n",
       "two full chains      3356\n",
       "no IR                  22\n",
       "Name: chain_pairing, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ir.tl.chain_pairing(adata_tcr)\n",
    "adata_tcr.obs['chain_pairing'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ed2d3a0",
   "metadata": {},
   "source": [
    "### 4.1. Mark Single Chains"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f70f533c",
   "metadata": {},
   "source": [
    "Generally, not each cell is assigned a full IR (VJ and VDJ part)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e70aaf2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\felix.drost\\Anaconda3\\envs\\covidVac\\lib\\site-packages\\ipykernel_launcher.py:1: FutureWarning: chain_pairing is a deprecated function and will be removed in a future version of scirpy. Use `tl.chain_qc` instead.\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "single pair        108395\n",
       "orphan VJ           38301\n",
       "ambiguous            4580\n",
       "extra VJ             3258\n",
       "orphan VDJ           2655\n",
       "two full chains       999\n",
       "extra VDJ             992\n",
       "no IR                 261\n",
       "multichain              5\n",
       "Name: chain_pairing, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ir.tl.chain_pairing(adata_bcr)\n",
    "adata_bcr.obs['chain_pairing'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7506d82c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bd703a74",
   "metadata": {},
   "source": [
    "### 4.2. Multi-Chains"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab55f469",
   "metadata": {},
   "source": [
    "Cells with more then 2 IRs are most likely dublets. Instead of sorting them out now, we will simply flag them and keep them in the dataset to use them as a marker later on. => we can also remove this dublet then from rna.\n",
    "However, amount of multi chains should be reasonable (X%) high amount of multi chains can indicate poor quality of the readouts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "97508ac2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    280045\n",
       "True          0\n",
       "None          0\n",
       "Name: multi_chain, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_tcr.obs['multi_chain'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c0be2ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:covidVac]",
   "language": "python",
   "name": "conda-env-covidVac-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
