{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "surprised-marking",
   "metadata": {},
   "source": [
    "# Introduction: Immune Receptor Profiling\n",
    "\n",
    "## Role IR in cells\n",
    "\n",
    "Immune Receptors (IR) are the key for the recognition of potential hazardous antigens and toxins invading the body. Those receptors are usually located in the membrane of speciallized cells, which recognize harmful content and initiate protective mechanisms. There are different kinds of IRs speciallized in the recognition of specific structures of foreign agents:\n",
    "\n",
    "- **Pattern recognition receptors (PRRs)**: Pathogen-associated molecular patterns (PAMPS)\n",
    "- **Killer activator/inhibitor receptors (KARs/KIRs)**: Host cells abnormalities\n",
    "- **Complement receptors**: Complement proteins\n",
    "- **Fc receptors**: Epitope-antibody complex\n",
    "- **Cytokine receptors**: Cytokines\n",
    "- **B-cell receptor (BCRs)**: Epitopes\n",
    "- **T-cell recpetors (TCRs)**: Linear epitopes bound to Major Histocompatibility Complex (MHC)\n",
    "\n",
    "\n",
    "## IRs of the  Adaptive Immune System\n",
    "In the adaptive immune system, there are two major linages of lymphocytes with IRs, which are produced in the Thymus and Bone-marrow called T- and B-cells {cite} 'cooper2006evolution'. These Adaptive Immune Receptors AIRs of both cell lines detect antigens derived from different pathogens and tumor cells. However, they interact with the antigens in different ways: While the B-cell receptor (BCR) directly recognizes soluble or membrane-bound epitopes, the T-cell receptor interacts with linear peptides bound to the surface protein MHC (pMHC), which enables to view inside cells by presenting its content on the cell's surface. After activation by antigens, B- and T-cells perform various functions such as fighting the pathogen, regulation of the immune response, or forming memory by proliferation.\n",
    "\n",
    "TODO: Image BCR / antibody, and TCR (discussion planned for 27.05.22)\n",
    "\n",
    "## Formation of AIRs by V(D)J Recombination\n",
    "\n",
    "TODO: Add Image of VDJ recombination (discussion planned for 27.05.22)\n",
    "\n",
    "The AIR is a protein complex consisting of two chains. Depending on the T cell type, these chains are either called α- and β- or γ- and δ-chain. Similarly, the BCR chains are called heavy and light chain (subtypes:  κ and λ). While the α-, γ-,  κ-, and λ-chains are formed by their respective V- and J-genes, the remaning chains additionally include their D-gene. For simplicity, we will therefore refer to VJ and VDJ chains throughout this book.\n",
    "\n",
    "Adaptive IRs lines are highly polymorphic in order to capture the immense space of antigens, i.e., there exists an immense variety of different TCR and BCR sequences. The amount of different TCRs have been estimated at 10<sup>20</sup> in total, while an human individual habours 10<sup>7</sup> different sequences {cite}'zarnitsyna2013estimating'. Similarly, there naturally occur up to 10<sup>18</sup> different BCRs {cite}'briney2019commonality'. The collection of all receptors forms the AIR repertoire of an individual. \n",
    "\n",
    "The sequence diversity is introduced by the V(D)J recombination mechanism:\n",
    "\n",
    "- **Combinatorial diversity** is introduced by the combination of different V-, (D-,) and J-genes in addition to the combination of different VJ and VDJ chains.\n",
    "- **Junctional diversity** is archieved by insertion of nucleotides at the gene interfaces.\n",
    "\n",
    "Additionally, mutations occur at the BCR binding sites in a process called Somatic Hypermutation for affinity maturation during rapid cell proliferation.\n",
    "\n",
    "In the resulting AIRs, three regions of high variability - the Complemenatary Determining Regions (CDRs) 1-3 - were detected per chain, where the AIR interacts with its target. The CDR1 and CDR2 are encoded by the V-gene. However, the CDR3 spans over the intersection of V-, (D-,) and J-genes and is therefore the most diverse element of the AIR. Therefore, it is often assumed that the specifity of an AIR is determined mainly by its CDR3 region of the more diverse VDJ chain.  \n",
    "    \n",
    "## VDJ-sequencing\n",
    "TODO: add explanation\n",
    "    - Single-cell isolation:\n",
    "    \n",
    "        - Flow-cytrometry\n",
    "        - Microfluidic devices\n",
    "        - Microwell plates\n",
    "        \n",
    "    - Single-cell library preparation:\n",
    "    \n",
    "        - Droplets with unique barcodes\n",
    "        - Emultion with magnetic beads {cite}'liu2021methods'\n",
    "        \n",
    "## AIR repertoire analysis\n",
    "\n",
    "\n",
    "VDJ-sequencing provides us with the nucleotide and thereby also the protein sequence of the AIR paired for both chains, from which the V-, (D-,) J-, and C-gene is determined in addition to the CDR3 sequence. Overall, the AIR sequence determines the specificity of the individual B- and T-cell. Therefore, the information obtained by VDJ-sequencing provides us with an indicator of the cells functionality, which is directly coupled to the AIRs target antigen. This enables us to use the AIR information in three major ways:\n",
    "- **Phenotyping**: We can group immune cells by identifying cells with the same or similar AIR, which share the same specificity. Having these groups, we can now observe, how disease-specific cells react under different conditions (e.g. transcriptomic change upon stimulation), whether immune cells have proliferated, or how the diversity of an immune repertoire changes upon after an immune response.\n",
    "- **Sequence Analysis**: Having identified groups of AIRs (e.g. a reactive cluster detected in other modalities), we can extract  properties of their sequence, such as V-, D-, and J-, gene usage or enriched sequence motifs, that are related to specific diseases or therapies.\n",
    "- **Specificity-Inference**: Last, we can use the sequence to match AIRs to their target antigen via database queries, sequence distances, or predictors. This directly identifies cells reactive to specific infectious diseases, tumors, or self-antigens. \n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acquired-nancy",
   "metadata": {},
   "source": [
    "# Dataset\n",
    "To showcase different approaches for preprocessing and analysis of IRs, we will use the dataset published by the Haniffa Lab {cite}'stephenson2021single', which contains transcriptome sequencing for over 750,000 Peripheral Blood Mononuclear Cells (PBMC) from 130 patients in the context of Severe Acute Respiratory Syndrome Coronavirus 2 (SARS-CoV-2).\n",
    "\n",
    "It consists of patients from three sources (Newcastle, Cambridge, and London) with different degrees of severity (assymptomatic, mild, mederate, severe, and critical) as well as negative controls (healthy, other severe respiratory illnesses, and healthy with Intravenous Liposaccharide to mimic systemic inflammatory response). Additional, patient-level information such as age, gender, and smoker status are provided. \n",
    "\n",
    "The dataset was choosen as an example of a large-scale single cell dataset including VDJ-Sequencing, that is well known to the community. It contains over 150,000 B-Cells and 200,000 T-Cells with IR annotation, that we will regard in this analysis tutorial."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "negative-wireless",
   "metadata": {},
   "source": [
    "To download the data please run the following commands:\n",
    "\n",
    "```\n",
    "wget https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-10026/E-MTAB-10026.processed.1.zip\n",
    "unzip E-MTAB-10026.processed.1.zip\n",
    "\n",
    "wget https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-10026/E-MTAB-10026.processed.2.zip\n",
    "unzip E-MTAB-10026.processed.2.zip\n",
    "```\n",
    "\n",
    "Or download and unzip from:\n",
    "- https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-10026/E-MTAB-10026.processed.1.zip\n",
    "- https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-10026/E-MTAB-10026.processed.2.zip\n",
    "\n",
    "#todo sent lukas files once decided on filtering => add to figshare"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "naughty-aging",
   "metadata": {},
   "source": [
    "# Load data\n",
    "In this tutorial we will mainly use two python packages for loading, cell-level ordering, and visualiation:\n",
    "- **Scanpy**: general package for single cell analysis (https://github.com/theislab/scanpy, {cite}'wolf2018scanpy')\n",
    "- **Scirpy**: scanpy extension for immune receptor analysis (https://github.com/scverse/scirpy, {cite}'sturm2020scirpy')\n",
    "\n",
    "Here, we only showcase IR analysis with scirpy. However, there exist several different tools with similar functionality such as immunarch(R, {cite}'immunomind2019'), scRepertoire (R, {cite}'borcherding2020screpertoire'), and dandelion (R, {cite}'stephenson2021single'), and Platypus (R, {cite}'yermanos2021platypus') reviewed in {cite}'valkiers2022recent'. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "elect-housing",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore', '.*IProgress not found*', )\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "import scanpy as sc\n",
    "import scirpy as ir\n",
    "\n",
    "import pandas as pd\n",
    "warnings.simplefilter(action='ignore', category=pd.errors.DtypeWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prime-acceptance",
   "metadata": {},
   "source": [
    "Let's set the input and output paths of our data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fifth-mercury",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_data = 'data/'\n",
    "\n",
    "path_bcr_input = f'{path_data}/scbcr_cellranger_filtered_contig_annotations.csv.gz'\n",
    "path_bcr_out = f'{path_data}/BCR_01_preprocessed.h5ad'\n",
    "\n",
    "path_tcr_input = f'{path_data}/TCR_merged-Updated.tsv'\n",
    "path_tcr_csv = f'{path_data}/TCR_merged-Updated.csv'\n",
    "path_tcr_out = f'{path_data}/TCR_01_preprocessed.h5ad'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "computational-quantum",
   "metadata": {},
   "source": [
    "## Raw data\n",
    "We beginn by with viewing the raw output of the cell ranger pipeline for a better understanding of the data we are working with.\n",
    "We will load the \"filtered_contig_annotations.csv\"-File to view its content. Each row will represent one measurement of a sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "urban-click",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total measurements: 373670\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>newBarcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGCGATATA-MH9143270</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGCGATATA-MH9143270_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>646</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV1-44</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ1</td>\n",
       "      <td>IGLC2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CAAWDDSLSASYVF</td>\n",
       "      <td>TGTGCAGCATGGGATGACAGCCTGAGTGCCTCTTATGTCTTC</td>\n",
       "      <td>154</td>\n",
       "      <td>14</td>\n",
       "      <td>clonotype9</td>\n",
       "      <td>clonotype9_consensus_2</td>\n",
       "      <td>COVID-002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGCGATATA-MH9143270</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGCGATATA-MH9143270_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>579</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-33</td>\n",
       "      <td>IGHD3-3</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARDNRVYYDFWSGYPDYW</td>\n",
       "      <td>TGTGCGAGAGACAACCGAGTCTATTACGATTTTTGGAGTGGTTATC...</td>\n",
       "      <td>144</td>\n",
       "      <td>13</td>\n",
       "      <td>clonotype9</td>\n",
       "      <td>clonotype9_consensus_1</td>\n",
       "      <td>COVID-002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGTCTACCAGA-MH9143270</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGTCTACCAGA-MH9143270_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>574</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV1-69D</td>\n",
       "      <td>IGHD2-15</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARDRRSAYCSGGSCWGGDWFDPW</td>\n",
       "      <td>TGTGCGAGAGATCGTCGTTCAGCTTATTGTAGTGGTGGTAGCTGCT...</td>\n",
       "      <td>564</td>\n",
       "      <td>42</td>\n",
       "      <td>clonotype10</td>\n",
       "      <td>clonotype10_consensus_1</td>\n",
       "      <td>COVID-002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGTCTACCAGA-MH9143270</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGTCTACCAGA-MH9143270_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>744</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV1-51</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ3</td>\n",
       "      <td>IGLC3</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CGTWDSSLSAWVF</td>\n",
       "      <td>TGCGGAACATGGGATAGCAGCCTGAGTGCTTGGGTGTTC</td>\n",
       "      <td>188</td>\n",
       "      <td>21</td>\n",
       "      <td>clonotype10</td>\n",
       "      <td>None</td>\n",
       "      <td>COVID-002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGGGCATTTGCTT-MH9143270</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACGGGCATTTGCTT-MH9143270_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>565</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-33</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CANPTRPYSSSWWYFDYW</td>\n",
       "      <td>TGTGCGAACCCCACCCGTCCATATAGCAGCAGCTGGTGGTACTTTG...</td>\n",
       "      <td>234</td>\n",
       "      <td>20</td>\n",
       "      <td>clonotype11</td>\n",
       "      <td>clonotype11_consensus_1</td>\n",
       "      <td>COVID-002</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      barcode  is_cell                            contig_id  \\\n",
       "0  AAACCTGAGCGATATA-MH9143270     True  AAACCTGAGCGATATA-MH9143270_contig_1   \n",
       "1  AAACCTGAGCGATATA-MH9143270     True  AAACCTGAGCGATATA-MH9143270_contig_2   \n",
       "2  AAACCTGTCTACCAGA-MH9143270     True  AAACCTGTCTACCAGA-MH9143270_contig_1   \n",
       "3  AAACCTGTCTACCAGA-MH9143270     True  AAACCTGTCTACCAGA-MH9143270_contig_2   \n",
       "4  AAACGGGCATTTGCTT-MH9143270     True  AAACGGGCATTTGCTT-MH9143270_contig_1   \n",
       "\n",
       "   high_confidence  length chain     v_gene    d_gene j_gene c_gene  \\\n",
       "0             True     646   IGL   IGLV1-44      None  IGLJ1  IGLC2   \n",
       "1             True     579   IGH   IGHV3-33   IGHD3-3  IGHJ4   IGHM   \n",
       "2             True     574   IGH  IGHV1-69D  IGHD2-15  IGHJ5   IGHM   \n",
       "3             True     744   IGL   IGLV1-51      None  IGLJ3  IGLC3   \n",
       "4             True     565   IGH   IGHV3-33      None  IGHJ4   IGHM   \n",
       "\n",
       "   full_length productive                      cdr3  \\\n",
       "0         True       True            CAAWDDSLSASYVF   \n",
       "1         True       True       CARDNRVYYDFWSGYPDYW   \n",
       "2         True       True  CARDRRSAYCSGGSCWGGDWFDPW   \n",
       "3         True      False             CGTWDSSLSAWVF   \n",
       "4         True       True        CANPTRPYSSSWWYFDYW   \n",
       "\n",
       "                                             cdr3_nt  reads  umis  \\\n",
       "0         TGTGCAGCATGGGATGACAGCCTGAGTGCCTCTTATGTCTTC    154    14   \n",
       "1  TGTGCGAGAGACAACCGAGTCTATTACGATTTTTGGAGTGGTTATC...    144    13   \n",
       "2  TGTGCGAGAGATCGTCGTTCAGCTTATTGTAGTGGTGGTAGCTGCT...    564    42   \n",
       "3            TGCGGAACATGGGATAGCAGCCTGAGTGCTTGGGTGTTC    188    21   \n",
       "4  TGTGCGAACCCCACCCGTCCATATAGCAGCAGCTGGTGGTACTTTG...    234    20   \n",
       "\n",
       "  raw_clonotype_id         raw_consensus_id patient_id newBarcode  \n",
       "0       clonotype9   clonotype9_consensus_2  COVID-002        NaN  \n",
       "1       clonotype9   clonotype9_consensus_1  COVID-002        NaN  \n",
       "2      clonotype10  clonotype10_consensus_1  COVID-002        NaN  \n",
       "3      clonotype10                     None  COVID-002        NaN  \n",
       "4      clonotype11  clonotype11_consensus_1  COVID-002        NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bcr_raw = pd.read_csv(path_bcr_input, index_col=0)\n",
    "\n",
    "# The column 'productive' contains mixed data types which are not compatible with downstream tools. \n",
    "# We correct them by casting them to strings.\n",
    "df_bcr_raw['productive'] = df_bcr_raw['productive'].astype(str)\n",
    "print(f'Total measurements: {len(df_bcr_raw)}')\n",
    "df_bcr_raw.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "burning-internship",
   "metadata": {},
   "source": [
    "The Table contains the following entries, that will be relevant for our preprocessing or analysis. A more detailed explanation of cellranger specific entries can be found here (https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation). In other data formats entries (e.g. AIRR) will have similiar, deviating names, however the underlying information remains similiar.\n",
    "- **barcode**: tag of the cell the contig was measured from\n",
    "- **is_cell**: indicates whether the barcode is associated with a cell\n",
    "- **high_confidence**: confidence of the measurement being a IR\n",
    "- **chain**: chain of the IR (e.g. TRA: T Cell Receptor α-chain, IGH: Immuneglobulin Heavy chain)\n",
    "- **{v,d,j,c}_gene**: gene used to form the specific segment of the IR\n",
    "- **full_length**: whether the full IR was captured (see below)\n",
    "- **productive**: whether the IR is productive (see below)\n",
    "- **cdr3{_nt}**: {Nucleotide} sequence of the CDR3 of the IR chain\n",
    "- **patient_id**: ID indicating the different patients"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "systematic-receipt",
   "metadata": {},
   "source": [
    "### Productive AIRs\n",
    "Even though we can detect the AIR sequence, it might not be productive, i.e., it might not form a valid AIR. Sequences, which do not result in functional AIRs, are therefore flagged as non-productive. These are usually ignored, when loading data by tools such as Scirpy, and not used for any downstream analysis.\n",
    "Productive Immune receptors are defined by 10x Genomics [here](https://kb.10xgenomics.com/hc/en-us/articles/115003248383-What-are-productive-contigs-) as:\n",
    "- Sequences spanning over from a V gene to a J-gene\n",
    "- Having a start codon in the leading region\n",
    "- Containing a CDR3 in the frame of the start codon.\n",
    "- Do not contain a stop codon within the V-J span\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "tender-washer",
   "metadata": {},
   "source": [
    "Example 1: Sequences not spanning over a full IR (V to J gene) as indicated by the 'full_length' column. You can find missing gene annotation here especially for V and J genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "advisory-pearl",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>productive</th>\n",
       "      <th>full_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>AACTCCCCAAGTACCT-MH9143270</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLC3</td>\n",
       "      <td>IGLJ5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>AACTCTTCAGATCGGA-MH9143270</td>\n",
       "      <td>IGKV1-39</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>AACTGGTCAGGGAGAG-MH9143270</td>\n",
       "      <td>IGLV4-3</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLC2</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>ACACCAAGTCGATTGT-MH9143270</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHD5-12</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ACATCAGCATACTACG-MH9143270</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHD3-3</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       barcode    v_gene    d_gene c_gene j_gene productive  \\\n",
       "34  AACTCCCCAAGTACCT-MH9143270      None      None  IGLC3  IGLJ5      False   \n",
       "38  AACTCTTCAGATCGGA-MH9143270  IGKV1-39      None   IGKC   None      False   \n",
       "43  AACTGGTCAGGGAGAG-MH9143270   IGLV4-3      None  IGLC2   None      False   \n",
       "65  ACACCAAGTCGATTGT-MH9143270      None  IGHD5-12   IGHD  IGHJ4      False   \n",
       "76  ACATCAGCATACTACG-MH9143270      None   IGHD3-3   IGHD  IGHJ6      False   \n",
       "\n",
       "    full_length  \n",
       "34        False  \n",
       "38        False  \n",
       "43        False  \n",
       "65        False  \n",
       "76        False  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = ['barcode', 'v_gene', 'd_gene', 'c_gene', 'j_gene', 'productive', 'full_length']\n",
    "df_bcr_raw[~df_bcr_raw['full_length']][columns].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "advance-journal",
   "metadata": {},
   "source": [
    "Example 2: Contigs express full length but there is not identifieable CDR3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "designing-neighbor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>productive</th>\n",
       "      <th>full_length</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGTCTACCAGA-MH9143270</td>\n",
       "      <td>IGLV1-51</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLC3</td>\n",
       "      <td>IGLJ3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>CGTWDSSLSAWVF</td>\n",
       "      <td>TGCGGAACATGGGATAGCAGCCTGAGTGCTTGGGTGTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AAAGCAAGTCGAAAGC-MH9143270</td>\n",
       "      <td>IGHV5-51</td>\n",
       "      <td>IGHD2-15</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AACGTTGAGCTGCGAA-MH9143270</td>\n",
       "      <td>IGHV3-30</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>CESGRATITMIVVTDLDYW</td>\n",
       "      <td>TGCGAAAGCGGGCGGGCTACGATTACTATGATAGTAGTGACTGACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AACTCAGTCAAAGTAG-MH9143270</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>IGHD3-16</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>CAKAWIGLDSEIYYDYIWGSYRIDFDYW</td>\n",
       "      <td>TGTGCGAAAGCTTGGATCGGACTCGACTCCGAGATTTATTATGATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>AACTCCCAGGTGCACA-MH9143270</td>\n",
       "      <td>IGKV6-21</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>IGKJ3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       barcode    v_gene    d_gene c_gene j_gene productive  \\\n",
       "3   AAACCTGTCTACCAGA-MH9143270  IGLV1-51      None  IGLC3  IGLJ3      False   \n",
       "13  AAAGCAAGTCGAAAGC-MH9143270  IGHV5-51  IGHD2-15   IGHD  IGHJ4      False   \n",
       "23  AACGTTGAGCTGCGAA-MH9143270  IGHV3-30      None   IGHM  IGHJ4      False   \n",
       "27  AACTCAGTCAAAGTAG-MH9143270  IGHV3-23  IGHD3-16   IGHD  IGHJ4      False   \n",
       "30  AACTCCCAGGTGCACA-MH9143270  IGKV6-21      None   IGKC  IGKJ3      False   \n",
       "\n",
       "    full_length                          cdr3  \\\n",
       "3          True                 CGTWDSSLSAWVF   \n",
       "13         True                          None   \n",
       "23         True           CESGRATITMIVVTDLDYW   \n",
       "27         True  CAKAWIGLDSEIYYDYIWGSYRIDFDYW   \n",
       "30         True                          None   \n",
       "\n",
       "                                              cdr3_nt  \n",
       "3             TGCGGAACATGGGATAGCAGCCTGAGTGCTTGGGTGTTC  \n",
       "13                                               None  \n",
       "23  TGCGAAAGCGGGCGGGCTACGATTACTATGATAGTAGTGACTGACC...  \n",
       "27  TGTGCGAAAGCTTGGATCGGACTCGACTCCGAGATTTATTATGATT...  \n",
       "30                                               None  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns += ['cdr3', 'cdr3_nt']\n",
    "df_bcr_raw[(df_bcr_raw['productive']=='False')\n",
    "           & (df_bcr_raw['full_length'])][columns].head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "needed-religious",
   "metadata": {},
   "source": [
    "We will now load the TCR data and convert it from tab-separated format to comma-separated format for later use. Since barcodes overlap between multiple samples within the dataset, we further copy the unique 'CellID' as 'barcode' annotation. This data also contains annotation on patient-level (e.g. Age, outcome, ...) and cell-level annotation (cell type) in addition to the contig annotations previously discussed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "widespread-economy",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total measurements: 547630\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Centre</th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>...</th>\n",
       "      <th>Days_from_onset</th>\n",
       "      <th>time_after_LPS</th>\n",
       "      <th>Worst_Clinical_Status</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>leiden</th>\n",
       "      <th>consensus</th>\n",
       "      <th>initial_clustering</th>\n",
       "      <th>study_id</th>\n",
       "      <th>AgeRange</th>\n",
       "      <th>Age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>AAACCTGAGAAACCTA-MH9179824</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAAACCTA-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>512</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV5-1</td>\n",
       "      <td>None</td>\n",
       "      <td>TRBJ2-3</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>26</td>\n",
       "      <td>CD8</td>\n",
       "      <td>CD8</td>\n",
       "      <td>COVID-014</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>AAACCTGAGAAACCTA-MH9179824</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAAACCTA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>573</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV8-4</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAJ31</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>26</td>\n",
       "      <td>CD8</td>\n",
       "      <td>CD8</td>\n",
       "      <td>COVID-014</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>AAACCTGAGAACAATC-MH9143324</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAACAATC-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>635</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV19</td>\n",
       "      <td>None</td>\n",
       "      <td>TRBJ1-5</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>25</td>\n",
       "      <td>gdT</td>\n",
       "      <td>gdT</td>\n",
       "      <td>COVID-029</td>\n",
       "      <td>(20, 29]</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>AAACCTGAGAACTCGG-MH8919329</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAACTCGG-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>499</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV34</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAJ30</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Death</td>\n",
       "      <td>Death</td>\n",
       "      <td>23</td>\n",
       "      <td>Treg</td>\n",
       "      <td>Treg</td>\n",
       "      <td>COVID-045</td>\n",
       "      <td>(70, 79]</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>AAACCTGAGAACTCGG-MH8919329</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAACTCGG-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>496</td>\n",
       "      <td>TRB</td>\n",
       "      <td>TRBV5-1</td>\n",
       "      <td>None</td>\n",
       "      <td>TRBJ2-5</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Death</td>\n",
       "      <td>Death</td>\n",
       "      <td>23</td>\n",
       "      <td>Treg</td>\n",
       "      <td>Treg</td>\n",
       "      <td>COVID-045</td>\n",
       "      <td>(70, 79]</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Centre                     barcode  is_cell  \\\n",
       "0  Newcastle  AAACCTGAGAAACCTA-MH9179824     True   \n",
       "1  Newcastle  AAACCTGAGAAACCTA-MH9179824     True   \n",
       "2  Newcastle  AAACCTGAGAACAATC-MH9143324     True   \n",
       "3  Newcastle  AAACCTGAGAACTCGG-MH8919329     True   \n",
       "4  Newcastle  AAACCTGAGAACTCGG-MH8919329     True   \n",
       "\n",
       "                     contig_id  high_confidence  length chain   v_gene d_gene  \\\n",
       "0  AAACCTGAGAAACCTA-1_contig_2             True     512   TRB  TRBV5-1   None   \n",
       "1  AAACCTGAGAAACCTA-1_contig_1             True     573   TRA  TRAV8-4   None   \n",
       "2  AAACCTGAGAACAATC-1_contig_1             True     635   TRB   TRBV19   None   \n",
       "3  AAACCTGAGAACTCGG-1_contig_1             True     499   TRA   TRAV34   None   \n",
       "4  AAACCTGAGAACTCGG-1_contig_2             True     496   TRB  TRBV5-1   None   \n",
       "\n",
       "    j_gene  ... Days_from_onset  time_after_LPS  Worst_Clinical_Status  \\\n",
       "0  TRBJ2-3  ...               8             NaN               Moderate   \n",
       "1   TRAJ31  ...               8             NaN               Moderate   \n",
       "2  TRBJ1-5  ...              16             NaN               Moderate   \n",
       "3   TRAJ30  ...               5             NaN                  Death   \n",
       "4  TRBJ2-5  ...               5             NaN                  Death   \n",
       "\n",
       "  Outcome leiden  consensus  initial_clustering   study_id  AgeRange Age  \n",
       "0    Home     26        CD8                 CD8  COVID-014  (50, 59]  54  \n",
       "1    Home     26        CD8                 CD8  COVID-014  (50, 59]  54  \n",
       "2    Home     25        gdT                 gdT  COVID-029  (20, 29]  26  \n",
       "3   Death     23       Treg                Treg  COVID-045  (70, 79]  77  \n",
       "4   Death     23       Treg                Treg  COVID-045  (70, 79]  77  \n",
       "\n",
       "[5 rows x 47 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcr_raw = pd.read_csv(path_tcr_input, sep='\\t')\n",
    "df_tcr_raw['barcode'] = df_tcr_raw.pop('CellID')  \n",
    "df_tcr_raw.to_csv(path_tcr_csv)\n",
    "print(f'Total measurements: {len(df_tcr_raw)}')\n",
    "df_tcr_raw.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "utility-audit",
   "metadata": {},
   "source": [
    "We will find that this data already has been processed: All contigs are of full length and represent a productive IR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "historical-penetration",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True    547630\n",
       "Name: full_length, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcr_raw['full_length'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "chemical-sheep",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True    547630\n",
       "Name: productive, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcr_raw['productive'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dominican-action",
   "metadata": {},
   "source": [
    "## Cell-Aligned Data\n",
    "In order to perform analysis on cell level, we next need to align IR information previously sorted by contig to a cell based format via the cell barcode. Note that usually many contigs map to the same cell, since every IR consists of VJ and VDJ chain. Also it has been shown, that a cell can express multiple IRs [cite].\n",
    "\n",
    "\n",
    "We will utilize the python package scirpy, that performs the cell-allignment automatically, when reading various TCR information from various formats:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "authentic-venezuela",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: Non-standard locus name ignored: Multi \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amount cells: 280045\n"
     ]
    }
   ],
   "source": [
    "adata_tcr = ir.io.read_10x_vdj(path_tcr_csv)\n",
    "print(f'Amount cells: {len(adata_tcr)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sound-engineering",
   "metadata": {},
   "source": [
    "The IR information will by stored sorted in adata_tcr.obs as a DataFrame with the index representing an individual cell. The DataFrame contains the following columns:\n",
    "- **has_ir**: indicates whether an IR was detected\n",
    "- **multi_chain**: indicates whether more than 2 IRs were detected\n",
    "- **extra_chains**: list of additional chains if multi_chain is true\n",
    "\n",
    "The contig data we observed in the raw data is stored as well. It is stored as seperate columns for the two chains (VJ and VDJ) with up to 2 contigs per chain for each cell:\n",
    "- IR_V{D}J_{1,2}_locus\n",
    "- IR_V{D}J_{1,2}_productive\n",
    "- IR_V{D}J_{1,2}_{v,d,j,c}_call\n",
    "- IR_V{D}J_{1,2}_junction{_aa}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "initial-jenny",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>is_cell</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>multi_chain</th>\n",
       "      <th>extra_chains</th>\n",
       "      <th>IR_VJ_1_c_call</th>\n",
       "      <th>IR_VJ_2_c_call</th>\n",
       "      <th>IR_VDJ_1_c_call</th>\n",
       "      <th>IR_VDJ_2_c_call</th>\n",
       "      <th>IR_VJ_1_consensus_count</th>\n",
       "      <th>IR_VJ_2_consensus_count</th>\n",
       "      <th>...</th>\n",
       "      <th>IR_VDJ_2_locus</th>\n",
       "      <th>IR_VJ_1_productive</th>\n",
       "      <th>IR_VJ_2_productive</th>\n",
       "      <th>IR_VDJ_1_productive</th>\n",
       "      <th>IR_VDJ_2_productive</th>\n",
       "      <th>IR_VJ_1_v_call</th>\n",
       "      <th>IR_VJ_2_v_call</th>\n",
       "      <th>IR_VDJ_1_v_call</th>\n",
       "      <th>IR_VDJ_2_v_call</th>\n",
       "      <th>has_ir</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACCTA-MH9179824</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>960.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAV8-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV5-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACAATC-MH9143324</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACTCGG-MH8919329</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5122.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAV34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV5-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGCCCA-MH9143373</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25914.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAV16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGACA-MH9143274</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3212.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>TRAV8-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBV7-6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           is_cell high_confidence multi_chain extra_chains  \\\n",
       "cell_id                                                                       \n",
       "AAACCTGAGAAACCTA-MH9179824    True            True       False           []   \n",
       "AAACCTGAGAACAATC-MH9143324    True            True       False           []   \n",
       "AAACCTGAGAACTCGG-MH8919329    True            True       False           []   \n",
       "AAACCTGAGAAGCCCA-MH9143373    True            True       False           []   \n",
       "AAACCTGAGAAGGACA-MH9143274    True            True       False           []   \n",
       "\n",
       "                           IR_VJ_1_c_call IR_VJ_2_c_call IR_VDJ_1_c_call  \\\n",
       "cell_id                                                                    \n",
       "AAACCTGAGAAACCTA-MH9179824           TRAC            NaN           TRBC2   \n",
       "AAACCTGAGAACAATC-MH9143324            NaN            NaN           TRBC1   \n",
       "AAACCTGAGAACTCGG-MH8919329           TRAC            NaN           TRBC2   \n",
       "AAACCTGAGAAGCCCA-MH9143373           TRAC            NaN             NaN   \n",
       "AAACCTGAGAAGGACA-MH9143274           TRAC            NaN           TRBC2   \n",
       "\n",
       "                           IR_VDJ_2_c_call  IR_VJ_1_consensus_count  \\\n",
       "cell_id                                                               \n",
       "AAACCTGAGAAACCTA-MH9179824             NaN                    960.0   \n",
       "AAACCTGAGAACAATC-MH9143324             NaN                      NaN   \n",
       "AAACCTGAGAACTCGG-MH8919329             NaN                   5122.0   \n",
       "AAACCTGAGAAGCCCA-MH9143373             NaN                  25914.0   \n",
       "AAACCTGAGAAGGACA-MH9143274             NaN                   3212.0   \n",
       "\n",
       "                            IR_VJ_2_consensus_count  ...  IR_VDJ_2_locus  \\\n",
       "cell_id                                              ...                   \n",
       "AAACCTGAGAAACCTA-MH9179824                      NaN  ...             NaN   \n",
       "AAACCTGAGAACAATC-MH9143324                      NaN  ...             NaN   \n",
       "AAACCTGAGAACTCGG-MH8919329                      NaN  ...             NaN   \n",
       "AAACCTGAGAAGCCCA-MH9143373                      NaN  ...             NaN   \n",
       "AAACCTGAGAAGGACA-MH9143274                      NaN  ...             NaN   \n",
       "\n",
       "                            IR_VJ_1_productive IR_VJ_2_productive  \\\n",
       "cell_id                                                             \n",
       "AAACCTGAGAAACCTA-MH9179824                True               None   \n",
       "AAACCTGAGAACAATC-MH9143324                None               None   \n",
       "AAACCTGAGAACTCGG-MH8919329                True               None   \n",
       "AAACCTGAGAAGCCCA-MH9143373                True               None   \n",
       "AAACCTGAGAAGGACA-MH9143274                True               None   \n",
       "\n",
       "                           IR_VDJ_1_productive IR_VDJ_2_productive  \\\n",
       "cell_id                                                              \n",
       "AAACCTGAGAAACCTA-MH9179824                True                None   \n",
       "AAACCTGAGAACAATC-MH9143324                True                None   \n",
       "AAACCTGAGAACTCGG-MH8919329                True                None   \n",
       "AAACCTGAGAAGCCCA-MH9143373                None                None   \n",
       "AAACCTGAGAAGGACA-MH9143274                True                None   \n",
       "\n",
       "                           IR_VJ_1_v_call  IR_VJ_2_v_call  IR_VDJ_1_v_call  \\\n",
       "cell_id                                                                      \n",
       "AAACCTGAGAAACCTA-MH9179824        TRAV8-4             NaN          TRBV5-1   \n",
       "AAACCTGAGAACAATC-MH9143324            NaN             NaN           TRBV19   \n",
       "AAACCTGAGAACTCGG-MH8919329         TRAV34             NaN          TRBV5-1   \n",
       "AAACCTGAGAAGCCCA-MH9143373         TRAV16             NaN              NaN   \n",
       "AAACCTGAGAAGGACA-MH9143274        TRAV8-1             NaN          TRBV7-6   \n",
       "\n",
       "                            IR_VDJ_2_v_call  has_ir  \n",
       "cell_id                                              \n",
       "AAACCTGAGAAACCTA-MH9179824              NaN    True  \n",
       "AAACCTGAGAACAATC-MH9143324              NaN    True  \n",
       "AAACCTGAGAACTCGG-MH8919329              NaN    True  \n",
       "AAACCTGAGAAGCCCA-MH9143373              NaN    True  \n",
       "AAACCTGAGAAGGACA-MH9143274              NaN    True  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_tcr.obs.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "statutory-transportation",
   "metadata": {},
   "source": [
    "Notice, that the patient-level information is not automatically added here. Let's add them by loading the raw data, alligning them on a cell level and indexing them by their barcode. \n",
    "\n",
    "TODO (aim 27.05.2022): discuss provide metadata as dict, or infer from data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "random-thread",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Centre</th>\n",
       "      <th>Sample</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>Collection_Day</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Swab_result</th>\n",
       "      <th>Status</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Status_on_day_collection</th>\n",
       "      <th>Status_on_day_collection_summary</th>\n",
       "      <th>Days_from_onset</th>\n",
       "      <th>time_after_LPS</th>\n",
       "      <th>Worst_Clinical_Status</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>initial_clustering</th>\n",
       "      <th>study_id</th>\n",
       "      <th>AgeRange</th>\n",
       "      <th>Age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACCTA-MH9179824</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>MH9179824</td>\n",
       "      <td>COVID-014</td>\n",
       "      <td>D0</td>\n",
       "      <td>Female</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Covid</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD8</td>\n",
       "      <td>COVID-014</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACAATC-MH9143324</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>MH9143324</td>\n",
       "      <td>COVID-029</td>\n",
       "      <td>D0</td>\n",
       "      <td>Male</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Covid</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>gdT</td>\n",
       "      <td>COVID-029</td>\n",
       "      <td>(20, 29]</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACTCGG-MH8919329</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>MH8919329</td>\n",
       "      <td>COVID-045</td>\n",
       "      <td>D0</td>\n",
       "      <td>Female</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Covid</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>ITU_intubated</td>\n",
       "      <td>Critical</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Death</td>\n",
       "      <td>Death</td>\n",
       "      <td>Treg</td>\n",
       "      <td>COVID-045</td>\n",
       "      <td>(70, 79]</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGCCCA-MH9143373</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>MH9143373</td>\n",
       "      <td>COVID-060</td>\n",
       "      <td>D0</td>\n",
       "      <td>Male</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Non_covid</td>\n",
       "      <td>Smoker</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Non_covid</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non-covid</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD4</td>\n",
       "      <td>COVID-060</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGACA-MH9143274</th>\n",
       "      <td>Newcastle</td>\n",
       "      <td>MH9143274</td>\n",
       "      <td>COVID-037</td>\n",
       "      <td>D0</td>\n",
       "      <td>Female</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Covid</td>\n",
       "      <td>Non-smoker</td>\n",
       "      <td>ITU_O2</td>\n",
       "      <td>Severe</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD4</td>\n",
       "      <td>COVID-037</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Centre     Sample patient_id Collection_Day  \\\n",
       "AAACCTGAGAAACCTA-MH9179824  Newcastle  MH9179824  COVID-014             D0   \n",
       "AAACCTGAGAACAATC-MH9143324  Newcastle  MH9143324  COVID-029             D0   \n",
       "AAACCTGAGAACTCGG-MH8919329  Newcastle  MH8919329  COVID-045             D0   \n",
       "AAACCTGAGAAGCCCA-MH9143373  Newcastle  MH9143373  COVID-060             D0   \n",
       "AAACCTGAGAAGGACA-MH9143274  Newcastle  MH9143274  COVID-037             D0   \n",
       "\n",
       "                               Sex Swab_result     Status      Smoker  \\\n",
       "AAACCTGAGAAACCTA-MH9179824  Female    Positive      Covid  Non-smoker   \n",
       "AAACCTGAGAACAATC-MH9143324    Male    Positive      Covid  Non-smoker   \n",
       "AAACCTGAGAACTCGG-MH8919329  Female    Positive      Covid  Non-smoker   \n",
       "AAACCTGAGAAGCCCA-MH9143373    Male    Negative  Non_covid      Smoker   \n",
       "AAACCTGAGAAGGACA-MH9143274  Female    Positive      Covid  Non-smoker   \n",
       "\n",
       "                           Status_on_day_collection  \\\n",
       "AAACCTGAGAAACCTA-MH9179824                  Ward_O2   \n",
       "AAACCTGAGAACAATC-MH9143324                  Ward_O2   \n",
       "AAACCTGAGAACTCGG-MH8919329            ITU_intubated   \n",
       "AAACCTGAGAAGCCCA-MH9143373                  Ward_O2   \n",
       "AAACCTGAGAAGGACA-MH9143274                   ITU_O2   \n",
       "\n",
       "                           Status_on_day_collection_summary Days_from_onset  \\\n",
       "AAACCTGAGAAACCTA-MH9179824                         Moderate               8   \n",
       "AAACCTGAGAACAATC-MH9143324                         Moderate              16   \n",
       "AAACCTGAGAACTCGG-MH8919329                         Critical               5   \n",
       "AAACCTGAGAAGCCCA-MH9143373                        Non_covid               3   \n",
       "AAACCTGAGAAGGACA-MH9143274                           Severe              21   \n",
       "\n",
       "                           time_after_LPS Worst_Clinical_Status Outcome  \\\n",
       "AAACCTGAGAAACCTA-MH9179824            NaN              Moderate    Home   \n",
       "AAACCTGAGAACAATC-MH9143324            NaN              Moderate    Home   \n",
       "AAACCTGAGAACTCGG-MH8919329            NaN                 Death   Death   \n",
       "AAACCTGAGAAGCCCA-MH9143373            NaN             Non-covid    Home   \n",
       "AAACCTGAGAAGGACA-MH9143274            NaN                Severe    Home   \n",
       "\n",
       "                           initial_clustering   study_id  AgeRange  Age  \n",
       "AAACCTGAGAAACCTA-MH9179824                CD8  COVID-014  (50, 59]   54  \n",
       "AAACCTGAGAACAATC-MH9143324                gdT  COVID-029  (20, 29]   26  \n",
       "AAACCTGAGAACTCGG-MH8919329               Treg  COVID-045  (70, 79]   77  \n",
       "AAACCTGAGAAGCCCA-MH9143373                CD4  COVID-060  (50, 59]   54  \n",
       "AAACCTGAGAAGGACA-MH9143274                CD4  COVID-037  (50, 59]   59  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_information = [\n",
    "    'barcode', 'Centre', 'Sample', 'patient_id', 'Collection_Day', 'Sex', \n",
    "    'Swab_result', 'Status', 'Smoker', 'Status_on_day_collection',\n",
    "    'Status_on_day_collection_summary', 'Days_from_onset', 'time_after_LPS',\n",
    "    'Worst_Clinical_Status', 'Outcome',\n",
    "    'initial_clustering', 'study_id', 'AgeRange', 'Age']\n",
    "df_patient = df_tcr_raw[patient_information].copy()\n",
    "df_patient['Days_from_onset'] = df_patient['Days_from_onset'].astype(str)  # mixed type (str, int)\n",
    "df_patient = df_patient.drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Assigning barcode as index\n",
    "df_patient.index = df_patient.pop('barcode')\n",
    "df_patient.index.name = None\n",
    "df_patient.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ranking-landscape",
   "metadata": {},
   "source": [
    "We can now add this information to the adata object and check, whether the annotation is added as observation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "distributed-silver",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>is_cell</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>multi_chain</th>\n",
       "      <th>extra_chains</th>\n",
       "      <th>IR_VJ_1_c_call</th>\n",
       "      <th>IR_VJ_2_c_call</th>\n",
       "      <th>IR_VDJ_1_c_call</th>\n",
       "      <th>IR_VDJ_2_c_call</th>\n",
       "      <th>IR_VJ_1_consensus_count</th>\n",
       "      <th>IR_VJ_2_consensus_count</th>\n",
       "      <th>...</th>\n",
       "      <th>Status_on_day_collection</th>\n",
       "      <th>Status_on_day_collection_summary</th>\n",
       "      <th>Days_from_onset</th>\n",
       "      <th>time_after_LPS</th>\n",
       "      <th>Worst_Clinical_Status</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>initial_clustering</th>\n",
       "      <th>study_id</th>\n",
       "      <th>AgeRange</th>\n",
       "      <th>Age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACCTA-MH9179824</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>960.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD8</td>\n",
       "      <td>COVID-014</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACAATC-MH9143324</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderate</td>\n",
       "      <td>Home</td>\n",
       "      <td>gdT</td>\n",
       "      <td>COVID-029</td>\n",
       "      <td>(20, 29]</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACTCGG-MH8919329</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5122.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ITU_intubated</td>\n",
       "      <td>Critical</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Death</td>\n",
       "      <td>Death</td>\n",
       "      <td>Treg</td>\n",
       "      <td>COVID-045</td>\n",
       "      <td>(70, 79]</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGCCCA-MH9143373</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25914.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Ward_O2</td>\n",
       "      <td>Non_covid</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non-covid</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD4</td>\n",
       "      <td>COVID-060</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGACA-MH9143274</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3212.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ITU_O2</td>\n",
       "      <td>Severe</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Severe</td>\n",
       "      <td>Home</td>\n",
       "      <td>CD4</td>\n",
       "      <td>COVID-037</td>\n",
       "      <td>(50, 59]</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           is_cell high_confidence multi_chain extra_chains  \\\n",
       "cell_id                                                                       \n",
       "AAACCTGAGAAACCTA-MH9179824    True            True       False           []   \n",
       "AAACCTGAGAACAATC-MH9143324    True            True       False           []   \n",
       "AAACCTGAGAACTCGG-MH8919329    True            True       False           []   \n",
       "AAACCTGAGAAGCCCA-MH9143373    True            True       False           []   \n",
       "AAACCTGAGAAGGACA-MH9143274    True            True       False           []   \n",
       "\n",
       "                           IR_VJ_1_c_call IR_VJ_2_c_call IR_VDJ_1_c_call  \\\n",
       "cell_id                                                                    \n",
       "AAACCTGAGAAACCTA-MH9179824           TRAC            NaN           TRBC2   \n",
       "AAACCTGAGAACAATC-MH9143324            NaN            NaN           TRBC1   \n",
       "AAACCTGAGAACTCGG-MH8919329           TRAC            NaN           TRBC2   \n",
       "AAACCTGAGAAGCCCA-MH9143373           TRAC            NaN             NaN   \n",
       "AAACCTGAGAAGGACA-MH9143274           TRAC            NaN           TRBC2   \n",
       "\n",
       "                           IR_VDJ_2_c_call  IR_VJ_1_consensus_count  \\\n",
       "cell_id                                                               \n",
       "AAACCTGAGAAACCTA-MH9179824             NaN                    960.0   \n",
       "AAACCTGAGAACAATC-MH9143324             NaN                      NaN   \n",
       "AAACCTGAGAACTCGG-MH8919329             NaN                   5122.0   \n",
       "AAACCTGAGAAGCCCA-MH9143373             NaN                  25914.0   \n",
       "AAACCTGAGAAGGACA-MH9143274             NaN                   3212.0   \n",
       "\n",
       "                            IR_VJ_2_consensus_count  ...  \\\n",
       "cell_id                                              ...   \n",
       "AAACCTGAGAAACCTA-MH9179824                      NaN  ...   \n",
       "AAACCTGAGAACAATC-MH9143324                      NaN  ...   \n",
       "AAACCTGAGAACTCGG-MH8919329                      NaN  ...   \n",
       "AAACCTGAGAAGCCCA-MH9143373                      NaN  ...   \n",
       "AAACCTGAGAAGGACA-MH9143274                      NaN  ...   \n",
       "\n",
       "                            Status_on_day_collection  \\\n",
       "cell_id                                                \n",
       "AAACCTGAGAAACCTA-MH9179824                   Ward_O2   \n",
       "AAACCTGAGAACAATC-MH9143324                   Ward_O2   \n",
       "AAACCTGAGAACTCGG-MH8919329             ITU_intubated   \n",
       "AAACCTGAGAAGCCCA-MH9143373                   Ward_O2   \n",
       "AAACCTGAGAAGGACA-MH9143274                    ITU_O2   \n",
       "\n",
       "                            Status_on_day_collection_summary Days_from_onset  \\\n",
       "cell_id                                                                        \n",
       "AAACCTGAGAAACCTA-MH9179824                          Moderate               8   \n",
       "AAACCTGAGAACAATC-MH9143324                          Moderate              16   \n",
       "AAACCTGAGAACTCGG-MH8919329                          Critical               5   \n",
       "AAACCTGAGAAGCCCA-MH9143373                         Non_covid               3   \n",
       "AAACCTGAGAAGGACA-MH9143274                            Severe              21   \n",
       "\n",
       "                           time_after_LPS Worst_Clinical_Status Outcome  \\\n",
       "cell_id                                                                   \n",
       "AAACCTGAGAAACCTA-MH9179824            NaN              Moderate    Home   \n",
       "AAACCTGAGAACAATC-MH9143324            NaN              Moderate    Home   \n",
       "AAACCTGAGAACTCGG-MH8919329            NaN                 Death   Death   \n",
       "AAACCTGAGAAGCCCA-MH9143373            NaN             Non-covid    Home   \n",
       "AAACCTGAGAAGGACA-MH9143274            NaN                Severe    Home   \n",
       "\n",
       "                            initial_clustering   study_id  AgeRange  Age  \n",
       "cell_id                                                                   \n",
       "AAACCTGAGAAACCTA-MH9179824                 CD8  COVID-014  (50, 59]   54  \n",
       "AAACCTGAGAACAATC-MH9143324                 gdT  COVID-029  (20, 29]   26  \n",
       "AAACCTGAGAACTCGG-MH8919329                Treg  COVID-045  (70, 79]   77  \n",
       "AAACCTGAGAAGCCCA-MH9143373                 CD4  COVID-060  (50, 59]   54  \n",
       "AAACCTGAGAAGGACA-MH9143274                 CD4  COVID-037  (50, 59]   59  \n",
       "\n",
       "[5 rows x 63 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_tcr.obs[df_patient.columns] = df_patient\n",
    "adata_tcr.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dying-container",
   "metadata": {},
   "source": [
    "In the same manner we can load the BCR data and check the IR annotation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "higher-routine",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: Non-standard locus name ignored: Multi \n",
      "WARNING: Non-standard locus name ignored: None \n"
     ]
    }
   ],
   "source": [
    "adata_bcr = ir.io.read_10x_vdj(path_bcr_input)\n",
    "adata_bcr.obs.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "brilliant-inventory",
   "metadata": {},
   "source": [
    "As before, we can add extract additional information annotated for each contig."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "executive-compromise",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_information = ['barcode', 'patient_id']\n",
    "df_patient = df_bcr_raw[patient_information].copy()\n",
    "df_patient = df_patient.drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Assigning barcode as index\n",
    "df_patient.index = df_patient.pop('barcode')\n",
    "df_patient.index.name = None\n",
    "df_patient.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "affecting-extent",
   "metadata": {},
   "source": [
    "And add them to the cell annotations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "nervous-guest",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_bcr.obs[df_patient.columns] = df_patient\n",
    "adata_bcr.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc8a8ab1",
   "metadata": {},
   "source": [
    "TODO downsampling for faster runs => depends on what to show in future chapters => not done yet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "educational-damages",
   "metadata": {},
   "source": [
    "# Quality Control"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "medium-practice",
   "metadata": {},
   "source": [
    "For analysis, we rely on high quality input data. It is therefore of great importance, to identify cells with incorrect or incomplete AIR information:\n",
    "- **Incomplete AIRs**: A cell is assigned only either a VJ and or a VDJ chain, because the other chain is missed during sequencing. While these are still valid cells, we cannot utilize them for downstream analysis, when full AIR sequence information is required.\n",
    "- **Multiple AIRs**: It is also possible that a cell gets assigned multiple AIRs. While it has been observed that T- / B-cells can express dual AIR {cite}'schuldt2019dual', cells with more than two IRs are indicative as doublets, which should not be used for downstream analysis.\n",
    "\n",
    "We can easily assign this state of the AIR by accumulating the data of the AIR chains. This will enable us later to filter cells, where the AIR cannot be used for our analysis. Here, it heavily depends on your downstream analysis, which AIR information is required."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6de1b369",
   "metadata": {},
   "source": [
    "## Assigning AIR State"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "focal-dubai",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ir.tl.chain_qc(adata_tcr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "continuing-situation",
   "metadata": {},
   "outputs": [],
   "source": [
    "ir.tl.chain_qc(adata_bcr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "blessed-baseline",
   "metadata": {},
   "source": [
    "The following provides the resulting column with an explanation of its possible options:\n",
    "\n",
    "- chain_pairing\n",
    "    - **orphan {VJ}/{VDJ}**: cells with only one of VJ or VDJ chain\n",
    "    - **single pair**: cells with a full paired (VJ and VDJ) AIR\n",
    "    - **extra {VJ}/{VDJ}**: cells with full paired AIR and an additional additional VJ or VDJ chain\n",
    "    - **two full chains**: cells with two paired AIRs\n",
    "    - **multichain**: cells with more than two VJ or VDJ chains (likely doublets)\n",
    "\n",
    "- receptor_type: BCR, TCR, no IR, or ambiguous (mixed chains of TCR and BCR)\n",
    "- receptor_subtype: α/β, ɣ/δ, IG-λ, and IG-κ chain configurations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6796008b",
   "metadata": {},
   "source": [
    "## Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "taken-exhaust",
   "metadata": {},
   "source": [
    "Depending on the data, the quality of chain_pairing might be different. You should visualize the results over different conditions (e.g. samples, data sources, ...), that could introduce varying quality within your dataset. Let's visualize the data  quality collected over the different centers of data collection)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "immediate-denial",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = ir.pl.group_abundance(adata_tcr, groupby='Centre', target_col='chain_pairing')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a192850d",
   "metadata": {},
   "source": [
    "The data from all three centers contains mainly cells with a single pair of TCRs, which is a sign of good data quality. Often, we can compare different conditions better, when we view the fractions of cells in each condition instead of the absolute amount of cells:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5721d23",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = ir.pl.group_abundance(adata_tcr, groupby='Centre', target_col='chain_pairing', normalize=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "524ca0c5",
   "metadata": {},
   "source": [
    "Here, we can see that all centers offer data of overall comparable data quality. The majority of T cells express a single TCR, followed by orphan VDJ, and VJ chains, which is typical when sequencing TCRs. \n",
    "Typically values of a good quality sample may be greater 60\\% single AIRs, 10-20\\% orphan chains, and ~10\\% extra chains.\n",
    "Note, that the percentage of cells without AIRs is naturally depending on the amount of non-AIR cells in the study, when sequencing multiple-modalities jointly.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "spectacular-winning",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = ir.pl.group_abundance(adata_tcr, groupby='Centre', target_col='receptor_type', normalize=True)\n",
    "_ = ir.pl.group_abundance(adata_tcr, groupby='Centre', target_col='receptor_subtype', normalize=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f068daae",
   "metadata": {},
   "source": [
    "We can see, that the dataset consists only of TCRs (first plot), which is not suprising, since we have seperated dataframes of T - and B-cells. Additionally, the dataset only contains α/β-T cells. Since the amount of cells without IR cannot be recognized in the plot, we will look at the absolute numbers, and observe that only 22 cells have no AIR information. This fraction will be larger in most datasets, when sequencing multiple modalities depending on the amount of AIR cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8f0daff",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_tcr.obs['chain_pairing'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "416cb4c8",
   "metadata": {},
   "source": [
    "For the BCR, we will have a look at the patient-level. For visualization, we downsample to two patients (todo: later in line with rest downsampling):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22ca3fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_bcr_tmp = adata_bcr[adata_bcr.obs['patient_id'].isin(['COVID-003', 'AP11'])].copy()\n",
    "_ = ir.pl.group_abundance(adata_bcr_tmp, groupby='patient_id', target_col='chain_pairing')\n",
    "_ = ir.pl.group_abundance(adata_bcr_tmp, groupby='patient_id', target_col='chain_pairing', normalize=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2213a10",
   "metadata": {},
   "source": [
    "While we observe an considerable difference in the amount of B Cells between the patients, the quality of the chain pairings is comparable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afb83388",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = ir.pl.group_abundance(adata_bcr_tmp, groupby='patient_id', target_col='receptor_type', normalize=True)\n",
    "_ = ir.pl.group_abundance(adata_bcr_tmp, groupby='patient_id', target_col='receptor_subtype', normalize=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71d704a9",
   "metadata": {},
   "source": [
    "Again, we have only one type of AIR in the dataset (BCRs), but this time a mixture of κ- and λ-chains for the light chain."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "heavy-excess",
   "metadata": {},
   "source": [
    "## Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pointed-terrain",
   "metadata": {},
   "source": [
    "You might want to filter cells with different AIR states. Depending on your study, you have to decide the trade-off between the amount of data available and the quality of the data, when deciding for filtering. Often, all cells are kept within the dataset with their AIR state flagged. For different downstream analysis, only cells with correct information are then used. E.g. if you perform a database query based on CDR3-VDJ chain (see chapter X), it is possible to use orphan VDJ chains, while this is not possible when searching for the full AIR. In the following, we will show how the filtering is performed. It is up to the reader to decide, whether they want to perform the filterin at this point, or before the different analysis parts."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35d5b4f2",
   "metadata": {},
   "source": [
    "The filtering can be performed with various degrees of strictness and different combinations such as:\n",
    "- **has IR**: keep only cells with expressed IRs, since we cannot perform sequence level analysis on them\n",
    "- **no multi chains**: filter out cells with >2 AIRs to avoid dublets\n",
    "- **only one AIR**: filter out cells with two AIRs or additional chains since specificity can not be attributed to a certain sequencev\n",
    "- **no single chains**: filter out cells with only VJ or VDJ chains, which cannot be used for all downstream analysis due to missing information\n",
    "\n",
    "Since we will use the full data throughout this book, we will filter the data to a temporary AnnData object for each step described above. From the absolute cell numbers, you can see how many cells will be lost in each step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "624faacb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_bcr_tmp.obs['chain_pairing'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eligible-midnight",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'Amount of all B cells:\\t\\t\\t\\t{len(adata_bcr)}')\n",
    "adata_bcr_tmp = adata_bcr[adata_bcr.obs['chain_pairing'] != 'no IR']\n",
    "print(f'Amount of B cells with AIR:\\t\\t\\t{len(adata_bcr_tmp)}')\n",
    "\n",
    "adata_bcr_tmp = adata_bcr_tmp[adata_bcr_tmp.obs['chain_pairing'] != 'multi_chain']\n",
    "print(f'Amount of B cells without dublets:\\t\\t{len(adata_bcr_tmp)}')\n",
    "\n",
    "adata_bcr_tmp = adata_bcr_tmp[~adata_bcr_tmp.obs['chain_pairing'].isin(['two full chains', 'extra VJ', 'extra VDJ'])]\n",
    "print(f'Amount of B cells with unique AIR per cell:\\t{len(adata_bcr_tmp)}')\n",
    "\n",
    "adata_bcr_tmp = adata_bcr_tmp[adata_bcr_tmp.obs['chain_pairing'] == 'single pair']\n",
    "print(f'Amount of B cells with sinlge complete AIR:\\t{len(adata_bcr_tmp)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d42ad42a",
   "metadata": {},
   "source": [
    "Eventually, we obtain an AnnData object with only single paired BCRs. However, the amount of data can be considerably smaller limiting downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74eb59b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_bcr_tmp.obs['chain_pairing'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "701b4207",
   "metadata": {},
   "source": [
    "Similarly, we can perform this filtering for the TCR data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d786f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_tcr.obs['chain_pairing'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b70d519c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'Amount of all T cells:\\t\\t\\t\\t{len(adata_tcr)}')\n",
    "adata_tcr_tmp = adata_tcr[adata_tcr.obs['chain_pairing'] != 'no IR']\n",
    "print(f'Amount of T cells with AIR:\\t\\t\\t{len(adata_tcr_tmp)}')\n",
    "\n",
    "adata_tcr_tmp = adata_tcr_tmp[adata_tcr_tmp.obs['chain_pairing'] != 'multi_chain']\n",
    "print(f'Amount of T cells without dublets:\\t\\t{len(adata_tcr_tmp)}')\n",
    "\n",
    "adata_tcr_tmp = adata_tcr_tmp[~adata_tcr_tmp.obs['chain_pairing'].isin(['two full chains', 'extra VJ', 'extra VDJ'])]\n",
    "print(f'Amount of T cells with unique AIR per cell:\\t{len(adata_tcr_tmp)}')\n",
    "\n",
    "adata_tcr_tmp = adata_tcr_tmp[adata_tcr_tmp.obs['chain_pairing'] == 'single pair']\n",
    "print(f'Amount of T cells with sinlge complete AIR:\\t{len(adata_tcr_tmp)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e48cdf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_tcr_tmp.obs['chain_pairing'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d2c644c",
   "metadata": {},
   "source": [
    "Finally, we will save the annotated data for later analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83f785ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.write(adata=adata_tcr, filename=path_tcr_out)\n",
    "sc.write(adata=adata_bcr, filename=path_bcr_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e99ba67",
   "metadata": {},
   "source": [
    "# Key Takeaways\n",
    "- VDJ-sequencing measures the sequence information of AIRs, which provides us insights into the function of B- and T-cells.\n",
    "- To detect doublets, cells with multiple AIRs should be identified or filtered.\n",
    "- Depending on the analysis, we additional filter cells with incomplete AIR information."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4367aa2",
   "metadata": {},
   "source": [
    "# Quiz\n",
    "Where does the high variability of AIRs stem from?\n",
    "- \\+ Juncitional diversity adds random nucleotides.\n",
    "- Peripheral diversity motivates rapid amino acid switches specific to the host organ.\n",
    "- \\+ Combinatorial diversity of different gene segments.\n",
    "- Inherited diversity genetically encodes multiple AIRs.\n",
    "\n",
    "What properties does a productive AIR have?\n",
    "- \\+ It spans over the V- to the J-gene with an CDR3 regions contained.\n",
    "- It contains a stop codon in the D-gene.\n",
    "- There exist multiple different sequences for the same AIR chains.\n",
    "- \\+ It contains a start codon in the leading region.\n",
    "\n",
    "What cells should be identified for potential filtering?\n",
    "- Cells with a high count of AIR sequence reads.\n",
    "- \\+ Cells with incomplete AIR information.\n",
    "- \\+ Cells with multiple AIRs indicating doublets.\n",
    "- Cells with specific motifs in the CDR3 region."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:best_practice]",
   "language": "python",
   "name": "conda-env-best_practice-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
