# Single-cell Analysis Tutorial

## Course presentation

```{admonition} Welcome!
:class: remove-from-content-only

The goal of this tutorial is to teach and to demonstrate best practices for the 
analysis of single-cell RNA-seq data. We take an objective view and do not focus on
particular tools from our labs or collaborators.
```

### Single-cell RNA-seq

Single-cell RNA-seq allows for the generation of expression profiles of individual cells at an unprecedented scale and resolution.
This information can for example be used to uncover rare cell types, to find stem cell development trajectories or to compare expression profiles of healthy and diseased patients. The inherent complexity of the such data has motivated the development of now more than 1000 tools {cite}`Zappia2021.08.13.456196`. The vast amounts of tools and the correct usage of these tools motivates a collection of recommendations for single-cell analysis best practices.

### Tutorial design

This tutorial is designed as an in-depth introduction to single-cell RNA-seq analysis best practices.
The lessons step-by-step introduce fundamental concepts and software tool allowing for a smooth learning experience into complex analysis patterns.
We emphasize that this tutorial is more than a cookbook: It will teach you how to choose the appropriate tools for the data, critically evaluate your analysis results and what biological conclusions to draw from them.

### Content and goals

The high level concepts that we cover are listed below. Note that we subdivided these concepts into subchapters if required.

1. Introduction to single-cell RNA-seq
2. Quality control
3. Normalization
4. Data integration
5. Dimensionality reduction
6. Visualization
7. Cluster analysis and annotation
8. Cell type composition analysis
9. Trajectory inference
10. Gene level analysis
11. Deconvolution
12. Latent space modeling & perturbations
13. Cell-cell interactions

## Prerequisites

The course aims to be accessible without a strong technical background. The
requirements for this course are:

- basic knowledge of Python programming : defining variables, writing
  functions, importing modules
- some prior experience with the NumPy, pandas and Matplotlib libraries is
  recommended but not required.

For a quick introduction on these requirements, you can go through these
[course materials](http://swcarpentry.github.io/python-novice-gapminder/)
or use the following resources:

- [Introduction to Python](https://scipy-lectures.org/intro/language/python_language.html)
- [Introduction to NumPy](https://sebastianraschka.com/blog/2020/numpy-intro.html)
- [Introduction to Pandas](https://pandas.pydata.org/docs/user_guide/10min.html)
- [Introduction to Matplotlib](https://sebastianraschka.com/blog/2020/numpy-intro.html#410-matplotlib)

LH ADD BASICS Of 
R
MOLECULAR BIOLOGY
SCANPY?!?

## Bibliography

```{bibliography}
```
